// @arcgis/map-components/dist/components/arcgis-track/customElement.js
/* COPYRIGHT Esri - https://js.arcgis.com/5.0/LICENSE.txt */
import { c as d } from "../../chunks/runtime.js";
import { css as g, html as e } from "lit";
import { usePropertyChange as v } from "@arcgis/lumina/controllers";
import { reEmitEvent as a } from "@arcgis/lumina/controllers/accessor";
import { classes as u } from "@arcgis/toolkit/dom";
import p from "@arcgis/core/widgets/Track/TrackViewModel.js";
import { LitElement as m, createEvent as f, safeClassMap as w, nothing as M } from "@arcgis/lumina";
import { u as b } from "../../chunks/useT9n.js";
import { m as k } from "../../chunks/useViewModel.js";
import { g as E } from "../../chunks/globalCss.js";
const T = g`.arcgis-button{--calcite-icon-color: var(--calcite-color-text-3);--calcite-button-background-color: var(--calcite-color-foreground-1);color:var(--calcite-color-text-3);line-height:var(--calcite-font-line-height-relative-tight);display:flex;flex-flow:column;cursor:pointer;background-color:var(--calcite-color-foreground-1);font-size:var(--calcite-font-size);:hover{--calcite-icon-color: var(--calcite-color-text-1);--calcite-button-background-color: var(--calcite-color-transparent-hover)}}:host{width:fit-content;height:fit-content}`, y = k(p);
class $ extends m {
  constructor() {
    super(...arguments), this.messages = b(), this.viewModel = y(this), this.view = this.viewModel.view, this.autoDestroyDisabled = !1, this.error = this.viewModel.error, this.geolocationOptions = this.viewModel.geolocationOptions, this.goToOverride = this.viewModel.goToOverride, this.goToLocationDisabled = this.viewModel.goToLocationEnabled, this.graphic = this.viewModel.graphic, this.icon = "compass-north-circle", this.positionFilterFunction = this.viewModel.positionFilterFunction, this.rotationDisabled = this.viewModel.rotationEnabled, this.scale = this.viewModel.scale, this.state = this.viewModel.state, this.tracking = this.viewModel.tracking, this.visualScale = "m", this.arcgisComplete = a(() => this.viewModel, "track"), this.arcgisError = a(() => this.viewModel, "track-error"), this.arcgisPropertyChange = v()("state"), this.arcgisReady = f();
  }
  static {
    this.properties = { view: 0, autoDestroyDisabled: 5, error: 0, geolocationOptions: 0, goToOverride: 0, goToLocationDisabled: 7, graphic: 0, icon: 1, label: 1, messageOverrides: 0, positionFilterFunction: 0, referenceElement: 1, rotationDisabled: 5, scale: 9, state: 0, tracking: 36, visualScale: 1 };
  }
  static {
    this.styles = T;
  }
  async destroy() {
    await this.manager.destroy();
  }
  async start() {
    this.viewModel.start();
  }
  async stop() {
    this.viewModel.stop();
  }
  _localizeError(t) {
    if (t instanceof GeolocationPositionError)
      return t.code === GeolocationPositionError.PERMISSION_DENIED ? this.messages?.permissionError : t.code === GeolocationPositionError.TIMEOUT ? this.messages?.timeoutError : this.messages?.positionUnavailable;
    if (t instanceof Error)
      return t.message;
  }
  _toggleTracking() {
    if (!this.viewModel)
      return;
    const { state: t } = this.viewModel;
    if (!(t === "feature-unsupported" || t === "disabled")) {
      if (t === "tracking" || t === "waiting") {
        this.viewModel.stop();
        return;
      }
      this.viewModel.start();
    }
  }
  render() {
    const { viewModel: t, messages: o, icon: c } = this, i = t?.state, l = i === "tracking", r = this._localizeError(this.viewModel?.error), s = (l ? o?.stopTracking : o?.startTracking) ?? "", n = i === "feature-unsupported", h = (i === "tracking" ? "pause" : void 0) ?? c ?? void 0;
    return e`<div class=${w(u("root", E.arcgisButton))}>${!n && e`<calcite-button .disabled=${i === "disabled"} .iconStart=${h} kind=neutral .label=${s} .loading=${i === "waiting"} @click=${this._toggleTracking} .scale=${this.visualScale} title=${s ?? M}></calcite-button>` || ""}${r ? e`<calcite-tooltip open overlay-positioning=fixed .referenceElement=${this.el}>${r}</calcite-tooltip>` : null}</div>`;
  }
}
d("arcgis-track", $);
export {
  $ as ArcgisTrack
};
