// @arcgis/map-components/dist/chunks/action-utils.js
/* COPYRIGHT Esri - https://js.arcgis.com/5.0/LICENSE.txt */
import "@arcgis/lumina";
import "lit/static-html.js";
import "@arcgis/core/intl.js";
import l from "@arcgis/core/support/actions/ActionButton.js";
import p from "@arcgis/core/support/actions/ActionToggle.js";
import { i as d } from "./async-utils.js";
import { a as m } from "./feature-utils.js";
import { isZoomScreenSize as b } from "@arcgis/core/applications/Components/actionUtils.js";
const n = new l({
  icon: "magnifying-glass-plus",
  id: "zoom-to-feature",
  title: "{messages.zoom}"
}), h = new l({
  icon: "trash",
  id: "remove-selected-feature",
  title: "{messages.remove}"
}), s = new l({
  icon: "magnifying-glass-plus",
  id: "zoom-to-clustered-features",
  title: "{messages.zoom}"
}), z = new p({
  icon: "table",
  id: "browse-clustered-features",
  title: "{messages.browseClusteredFeatures}",
  value: !1
});
function P(e) {
  if (e.icon)
    return e.icon;
  if (!("image" in e && e.image) && !e.className)
    return "question";
}
function Q(e) {
  return e ? {
    backgroundImage: `url(${e})`
  } : {};
}
function E(e) {
  const { selectedFeature: t, location: a, view: o } = e;
  return o ? t ?? a ?? null : null;
}
async function C(e, t) {
  const { location: a, selectedFeature: o, view: r, zoomFactor: i } = e;
  await t?.viewModel?.updateGeometry();
  const c = t?.graphic, g = c?.geometry ? c : E(e);
  if (!r || !g)
    throw new Error("Cannot zoom to location without a target and view.");
  const w = r.scale / i, u = c?.geometry ?? e.selectedFeature?.geometry ?? a, f = u?.type === "point" && await b(c ?? o, r);
  n.active = !0, n.disabled = !0;
  try {
    await e.zoomTo({
      target: {
        target: g,
        scale: f ? w : void 0
      }
    });
  } catch (y) {
    if (d(y))
      return;
    throw new Error("Could not zoom to the location of the graphic.");
  } finally {
    n.active = !1, n.disabled = !1, f && e.location !== u && (e.location = u);
  }
}
function v(e) {
  return !!e && e.isAggregate && m(e)?.featureReduction?.type === "cluster";
}
async function F(e, t) {
  const a = await e.whenLayerView(m(t)), o = a.createQuery(), r = t.getObjectId();
  return o.aggregateIds = r != null ? [r] : [], [a, o];
}
async function T(e) {
  const { selectedFeature: t, view: a } = e;
  if (a?.type !== "2d")
    throw new Error("zoomToCluster:invalid-view: View must be 2d MapView.");
  if (!t || !v(t))
    throw new Error("zoomToCluster:invalid-selectedFeature: Selected feature must represent an aggregate/cluster graphic.");
  const [o, r] = await F(a, t);
  s.active = !0, s.disabled = !0;
  const { extent: i } = await o.queryExtent(r);
  i && await e.zoomTo({ target: i }), s.active = !1, s.disabled = !1;
}
const Z = async (e, t) => {
  if (!e)
    throw new Error("trigger-action:missing-arguments: Event has no action");
  const { disabled: a, id: o } = e;
  if (!o)
    throw new Error("trigger-action:invalid-action: action.id is missing");
  if (a)
    throw new Error("trigger-action:invalid-action: Action is disabled");
  if (o === n.id)
    return await C(t).catch((r) => {
      if (!d(r))
        throw r;
    });
  if (o === s.id)
    return await T(t);
  if (o === z.id) {
    t.browseClusterEnabled = !t.browseClusterEnabled, t.featureMenuOpen = t.browseClusterEnabled;
    return;
  }
  if (o === h.id) {
    t.visible = !1;
    const { selectedFeature: r } = t;
    if (!r)
      throw new Error("trigger-action:missing-selected-feature: Cannot remove selected feature without a selected feature.");
    const i = m(r);
    i ? i.remove(r) : t.view?.graphics.remove(r);
  }
};
export {
  Q as a,
  C as b,
  P as g,
  Z as t,
  n as z
};
