// @arcgis/map-components/dist/chunks/component-utils.js
/* COPYRIGHT Esri - https://js.arcgis.com/5.0/LICENSE.txt */
import { watch as u } from "@arcgis/core/core/reactiveUtils.js";
import "@arcgis/toolkit/log";
function N(e) {
  e.el.childElem && (e.el.childElem.ownedBy = e.el);
  const t = g(e);
  if (!t) {
    const r = y(e);
    if (!r) {
      c(e);
      return;
    }
    return a(r, e, !0), r;
  }
  return E(t, e.el.parent) || (e.el.parent = t, t.localName === "arcgis-expand" ? h(t, e) : a(t, e, !1)), t;
}
function g({ el: e }) {
  for (let t = e.parentElement; t; t = t?.parentElement ?? null) {
    if (o.has(t.localName))
      return t;
    if ("ownedBy" in t && t.ownedBy instanceof HTMLElement && o.has(t.ownedBy.localName))
      return t.ownedBy;
  }
}
const o = /* @__PURE__ */ new Set(["arcgis-map", "arcgis-scene", "arcgis-link-chart", "arcgis-expand"]);
function E(e, t) {
  return e !== t ? !1 : e.localName === "arcgis-expand";
}
function s(e, t, r = "arcgisReady") {
  const i = e;
  if (i.view)
    t(i.view);
  else {
    let n = function(l) {
      l.target === e && (s(e, t, r), e.removeEventListener(r, n));
    };
    e.addEventListener(r, n);
  }
}
const h = (e, t) => s(e, (r) => {
  const i = d(t);
  if (t.el.shadowRoot ? e.content = t.el : t.el.childElem && e.content != null && i.append(t.el.childElem), t.el.view = r, e.referenceElement) {
    t.referenceElement = e.referenceElement;
    return;
  }
  u(
    () => e.referenceElement,
    (n) => {
      t.referenceElement = n;
    },
    {
      once: !0
    }
  );
}), a = (e, t, r) => s(
  e,
  (i) => {
    r && c(t), t.el.view = i, t.referenceElement = e;
  },
  "arcgisViewReadyChange"
);
function S(e) {
  return e == null || e === "";
}
function T(e, t) {
  return e?.position === t?.position && e?.heading === t.heading && e?.tilt === t.tilt && e?.fov === t.fov;
}
function B(e, t) {
  return t === void 0 || e?.targetGeometry !== t?.targetGeometry || e?.rotation !== t?.rotation || e?.scale !== t?.scale;
}
function m(e, t) {
  return f(t?.center, e);
}
function f(e, t) {
  if (typeof t == "string")
    return f(e, t.split(",").map(Number));
  if (!e)
    return t;
  if (Array.isArray(t)) {
    const r = t[0], i = t[1], n = t.length > 2 ? t[2] : void 0;
    if (r !== e.longitude || i !== e.latitude || n !== e.z)
      return t;
  } else if (t && !e?.equals(t))
    return t;
}
function c(e) {
  e.el.childElem && d(e).append(e.el.childElem);
}
function d(e) {
  return e.el.shadowRoot ?? e.el;
}
function y(e, t) {
  const r = e.referenceElement;
  if (typeof r != "string")
    return r ?? void 0;
  const i = r.includes("#") || r.includes(".") || r.includes("["), n = e.el.getRootNode();
  return (i ? void 0 : n.querySelector(`#${r}`)) ?? n.querySelector(r) ?? void 0;
}
const q = (e) => e.el.childElem;
async function F(e) {
  const t = typeof e == "function" ? e() : e;
  t && ("setFocus" in t && typeof t.setFocus == "function" ? await t.setFocus() : t instanceof HTMLElement && t.focus());
}
export {
  N as a,
  T as b,
  f as c,
  S as d,
  q as g,
  B as i,
  m as p,
  F as s
};
