// @arcgis/map-components/dist/chunks/feature-utils.js
/* COPYRIGHT Esri - https://js.arcgis.com/5.0/LICENSE.txt */
import { replace as a } from "@arcgis/core/applications/Components/stringUtils.js";
import { fieldFormatFromFieldInfo as m } from "@arcgis/core/applications/Components/fieldFormatUtils.js";
import { g as F } from "./layer-utils.js";
import { log as g } from "@arcgis/toolkit/log";
const u = /^\s*expression\//iu, y = /href=(""|'')/giu, C = /(\{([^{\r\n]+)\})/gu, L = /href\s*=\s*(?:"([^"]+)"|'([^']+)')/giu, h = /'/gu, N = "relationships/", b = /^(?:mailto:|tel:)/u;
function P(e, t) {
  const n = /* @__PURE__ */ new Map();
  if (!e)
    return n;
  for (const r of e) {
    if (!r.fieldName)
      continue;
    const i = c(r.fieldName, t);
    r.fieldName = i, n.set(i.toLowerCase(), r);
  }
  return n;
}
function U(e) {
  const t = [];
  if (!e)
    return t;
  const { fieldInfos: n, content: r } = e;
  return n && t.push(...n), !r || !Array.isArray(r) || r.forEach((i) => {
    if (i.type === "fields") {
      const o = i?.fieldInfos;
      o && t.push(...o);
    }
  }), t;
}
function $(e, t) {
  return e?.getEffectivePopupTemplate(t) ?? void 0;
}
function D(e) {
  if (e != null)
    return "sourceLayer" in e && e.sourceLayer ? e.sourceLayer : e.layer ?? void 0;
}
function I(e = "") {
  return e ? e.includes(N) : !1;
}
function W(e = "") {
  if (e)
    return !b.test(e.trim().toLowerCase());
}
function c(e, t) {
  const n = p(t, e);
  return n ? n.name : e;
}
function q(e, t) {
  return e?.map((n) => c(n, t));
}
function f(e) {
  return e ? u.test(e) : !1;
}
function w(e, t) {
  if (!t || !f(t) || !e)
    return;
  const n = t.replace(u, "").toLowerCase();
  return e.find(({ name: r }) => r.toLowerCase() === n);
}
function _(e, t) {
  if (!(!e?.length || !t))
    return e.find((n) => n.fieldName?.toLowerCase() === t.toLowerCase());
}
function A(e) {
  const { fieldInfo: t, graphic: n, layer: r } = e;
  return T({ fieldInfo: t, graphic: n }) ? F(r, { graphic: n }) : null;
}
function T({
  fieldInfo: e,
  graphic: t
}) {
  return !!e?.fieldName && !f(e.fieldName) && !I(e.fieldName) && !t?.popupTemplate;
}
function B({
  expressionInfos: e,
  fieldInfo: t,
  graphic: n,
  isContentFieldInfos: r,
  layer: i
}) {
  if (!t?.fieldName)
    return null;
  const o = w(e, t?.fieldName);
  if (o)
    return o.title || null;
  const s = A({ fieldInfo: t, graphic: n, layer: i });
  if (s) {
    const l = s.getFieldAlias(t.fieldName);
    return r ? t.label || l || t.fieldName : l || t.fieldName;
  }
  return t.label || t.fieldName;
}
function G(e) {
  const { configurableFieldsContainer: t, fieldInfo: n, isContentFieldInfos: r } = e;
  if (!n?.fieldName)
    return null;
  const i = !t.popupTemplate && !t.fieldConfigurations ? m(n, t.getField(n.fieldName)) : t.getFieldConfiguration(n.fieldName)?.fieldFormat;
  return r && n.fieldFormat || i;
}
function d(e) {
  return e.trim();
}
function x(e) {
  return e.replaceAll(y, "");
}
function k(e, t) {
  return `{${t.get(e.toLowerCase())?.fieldName || e}}`;
}
function R({
  formattedValues: e,
  template: t,
  fieldInfoMap: n
}) {
  return d(
    x(
      a(
        a(t, (r) => k(r, n)),
        e
      )
    )
  );
}
function p(e, t) {
  return !(e && "getField" in e && typeof e.getField == "function") || !t ? null : e.getField(t) ?? null;
}
function E(e, t) {
  return e.replaceAll(C, (n, r, i) => {
    const o = p(t, i);
    return o ? `{${o.name}}` : r;
  });
}
function V(e, t, n = !1) {
  const r = t[e];
  if (typeof r == "string") {
    const o = (n ? encodeURIComponent(r) : r).replaceAll(
      h,
      "%27"
    );
    t[e] = o;
  }
}
function v(e, t = !1) {
  const n = { ...e };
  return Object.keys(n).forEach((r) => V(r, n, t)), n;
}
function S(e, t, n) {
  t = d(t);
  const r = t && !t.startsWith("{");
  return a(e, v(n, r || !1));
}
function j(e, t, n) {
  const r = E(e, n);
  return r && r.replaceAll(
    L,
    (i, o, s) => S(i, o || s, t)
  );
}
function J({
  values: e,
  globalValues: t,
  layer: n,
  text: r,
  fieldInfoMap: i
}) {
  return r ? R({
    formattedValues: t,
    template: j(
      r,
      {
        ...t,
        ...e
      },
      n
    ),
    fieldInfoMap: i
  }) : "";
}
async function K({
  type: e,
  value: t,
  event: n
}) {
  try {
    return typeof t == "function" ? t(n) : await t;
  } catch (r) {
    g("error", "arcgis-feature", `An error occurred when calling the "${e}" function`, {
      detail: {
        error: r,
        graphic: n.graphic,
        value: t
      }
    });
    return;
  }
}
function Q(e) {
  const t = e.attributes, n = e.layer;
  if (!t || !n)
    return "";
  const r = "displayField" in n ? n.displayField : null, i = r != null ? t[r] : null, o = e.getObjectId()?.toString();
  return i?.toString() || o || "Unknown";
}
export {
  D as a,
  U as b,
  P as c,
  I as d,
  K as e,
  W as f,
  $ as g,
  Q as h,
  f as i,
  B as j,
  A as k,
  G as l,
  E as m,
  R as n,
  q as o,
  c as p,
  _ as q,
  J as s
};
