// @arcgis/map-components/dist/chunks/layer-utils.js
/* COPYRIGHT Esri - https://js.arcgis.com/5.0/LICENSE.txt */
function a(e, t) {
  if (!e.map)
    return;
  const { map: n } = e;
  n.layers.includes(t) || (n.add(t), n.reorder(t, n.layers.length - 1));
}
function o(e) {
  return e != null && typeof e == "object" && "type" in e && e.type === "feature";
}
function s(e) {
  return o(e) && e?.type === "feature" && e.source?.type === "feature-layer";
}
function f(e) {
  return e != null && typeof e == "object" && "type" in e && e.type === "graphics";
}
function p(e) {
  return e?.type === "subtype-sublayer";
}
function c(e) {
  return o(e) && s(e);
}
function d(e, t) {
  const n = t?.type === "subtype-sublayer" ? t.parent : t;
  return e.allLayerViews.find((r) => {
    const u = r.layer;
    return u === n || "sublayers" in u && u.sublayers != null && u.sublayers.includes(n);
  });
}
function i(e) {
  const t = e?.type;
  return t === "binning" || t === "cluster";
}
function y(e, t) {
  if (!c(e))
    return;
  const n = t?.graphic, r = e.featureReduction;
  if (n) {
    if (n.isAggregate) {
      const u = n.origin;
      return u && "featureReductionProvider" in u && r ? i(r) ? r : null : void 0;
    }
    return e;
  }
  return t?.checkFeatureReduction && r ? i(r) ? r : null : e;
}
export {
  a,
  p as b,
  d as f,
  y as g,
  f as i
};
