// @arcgis/map-components/dist/chunks/time-utils.js
/* COPYRIGHT Esri - https://js.arcgis.com/5.0/LICENSE.txt */
const l = "system", d = "unknown", f = Intl.DateTimeFormat().resolvedOptions().timeZone, o = {
  milliseconds: 1,
  seconds: 1e3,
  minutes: 6e4,
  hours: 36e5,
  days: 864e5,
  weeks: 6048e5,
  months: 26784e5,
  // 31 days
  years: 31536e6,
  // 365 days
  decades: 31536e7,
  centuries: 31536e8
};
function c(e, s, n) {
  return e === 0 ? 0 : e * o[s] / o[n];
}
function r(e) {
  return e instanceof Date && !Number.isNaN(e.getTime());
}
function g(e, s, n = "hours") {
  const t = new Date(e.getTime()), i = c(s, n, "milliseconds");
  return t.setTime(t.getTime() + i), t;
}
function u(e, s) {
  const n = c(s, "hours", "milliseconds"), t = new Date(e.getTime() + n), i = t.toISOString().slice(0, 10), a = t.getUTCHours() * 60 + t.getUTCMinutes();
  return { localDate: i, minutesSinceMidnight: a };
}
function T(e) {
  const s = e?.environment?.lighting;
  if (m(s)) {
    const n = s.displayUTCOffset ?? 0, t = s.date;
    if (r(t)) {
      const { localDate: i } = u(t, n);
      return i;
    }
  }
}
function m(e) {
  return e?.type === "sun";
}
export {
  f as a,
  r as b,
  c,
  u as d,
  T as g,
  m as i,
  g as o,
  l as s,
  d as u
};
