// @arcgis/map-components/dist/components/arcgis-popup/customElement.js
/* COPYRIGHT Esri - https://js.arcgis.com/5.0/LICENSE.txt */
import { c as A } from "../../chunks/runtime.js";
import { css as E, html as g } from "lit";
import { keyed as P } from "lit/directives/keyed.js";
import F from "@arcgis/core/widgets/Features/FeaturesViewModel.js";
import { u as $ } from "../../chunks/useT9n.js";
import { useDirection as N, usePropertyChange as O } from "@arcgis/lumina/controllers";
import { m as L } from "../../chunks/useViewModel.js";
import { LitElement as S, createEvent as _, safeStyleMap as T, safeClassMap as y } from "@arcgis/lumina";
import { stripHTML as V } from "@arcgis/core/applications/Components/stringUtils.js";
import { classes as C } from "@arcgis/toolkit/dom";
import { createRef as m, ref as f } from "lit/directives/ref.js";
import { watch as l } from "@arcgis/core/core/reactiveUtils.js";
import B from "@arcgis/core/support/actions/ActionButton.js";
import { t as H } from "../../chunks/screenLocation.js";
import "../arcgis-search-result-renderer/customElement.js";
const I = E`:host{display:contents;--arcgis-internal-popup-pointer-size: 12px;--arcgis-internal-popup-content-small-size: 280px;--arcgis-internal-popup-content-medium-size: 340px;--arcgis-internal-popup-content-large-size: 400px;--arcgis-internal-popup-content-x-large-size: 460px;--arcgis-internal-popup-content-min-size-y: 48px;--arcgis-internal-popup-pointer-size-half: calc(var(--arcgis-internal-popup-pointer-size) / 2);--arcgis-internal-popup-pointer-size-half-negative: calc(var(--arcgis-internal-popup-pointer-size-half) * -1)}.root{display:flex;position:absolute;flex-flow:column nowrap;z-index:401;pointer-events:none;border:none;overflow:visible;padding:0;width:auto;height:auto;background-color:transparent}.hidden{visibility:hidden}.root:popover-open{display:flex;position:fixed;flex-flow:column nowrap;margin:0}arcgis-features{flex:1;width:100%}.shadow{box-shadow:var(--calcite-shadow-sm)}.aligned-top-center{transform-origin:50% 100%}.aligned-bottom-center{transform-origin:50% -5%}.aligned-top-left,.aligned-bottom-left{transform-origin:100% 50%}.aligned-top-right,.aligned-bottom-right{transform-origin:-5% 50%}.aligned-top-center,.aligned-top-left,.aligned-top-right{animation:popup-intro-animation-down .2s ease-in-out}.aligned-bottom-center,.aligned-bottom-left,.aligned-bottom-right{animation:popup-intro-animation-up .2s ease-in-out}.main-container{display:flex;position:relative;z-index:1;background-color:var(--calcite-color-foreground-1);width:var(--arcgis-internal-popup-content-medium-size);min-height:var(--arcgis-internal-popup-content-min-size-y);max-height:var(--arcgis-internal-popup-content-medium-size);overflow:hidden;pointer-events:auto}.pointer{position:absolute;width:0;height:0}.pointer-direction{position:absolute;width:var(--arcgis-internal-popup-pointer-size);height:var(--arcgis-internal-popup-pointer-size);content:"";background-color:var(--calcite-color-foreground-1)}.aligned-top-center .pointer{top:100%;left:50%;margin:0 0 0 var(--arcgis-internal-popup-pointer-size-half-negative)}.aligned-bottom-center .pointer{bottom:100%;left:50%;margin:0 0 0 var(--arcgis-internal-popup-pointer-size-half-negative)}.aligned-top-left .pointer{right:var(--arcgis-internal-popup-pointer-size-half);bottom:var(--arcgis-internal-popup-pointer-size-half);transform:rotate(-45deg)}.aligned-bottom-left .pointer{top:var(--arcgis-internal-popup-pointer-size-half);right:var(--arcgis-internal-popup-pointer-size-half);transform:rotate(45deg)}.aligned-top-right .pointer{bottom:var(--arcgis-internal-popup-pointer-size-half);left:var(--arcgis-internal-popup-pointer-size-half);transform:rotate(45deg)}.aligned-bottom-right .pointer{top:var(--arcgis-internal-popup-pointer-size-half);left:var(--arcgis-internal-popup-pointer-size-half);transform:rotate(-45deg)}.aligned-top-center .pointer-direction,.aligned-bottom-center .pointer-direction{transform:scale(.75,2) rotate(45deg)}.aligned-top-left .pointer-direction,.aligned-bottom-left .pointer-direction,.aligned-top-right .pointer-direction,.aligned-bottom-right .pointer-direction{top:var(--arcgis-internal-popup-pointer-size-half-negative);left:var(--arcgis-internal-popup-pointer-size-half-negative);transform:scale(.75,2.5) rotate(45deg)}.aligned-top-center .pointer-direction{top:var(--arcgis-internal-popup-pointer-size-half-negative);left:0}.aligned-bottom-center .pointer-direction{bottom:var(--arcgis-internal-popup-pointer-size-half-negative);left:0}.is-xl-width-breakpoint .main-container{width:var(--arcgis-internal-popup-content-x-large-size)}.is-lg-width-breakpoint .main-container{width:var(--arcgis-internal-popup-content-large-size)}.is-md-width-breakpoint .main-container{width:var(--arcgis-internal-popup-content-medium-size)}.is-sm-width-breakpoint .main-container,.is-xs-width-breakpoint .main-container{width:var(--arcgis-internal-popup-content-small-size)}.is-xs-width-breakpoint.is-docked-top-center,.is-xs-width-breakpoint.is-docked-bottom-center{margin:0;width:auto}.is-xs-width-breakpoint.is-docked-top-center .main-container,.is-xs-width-breakpoint.is-docked-bottom-center .main-container{position:absolute;right:0;left:0;width:100%;max-height:75%}.is-xs-width-breakpoint.is-docked-bottom-center .main-container{bottom:0}.is-xs-width-breakpoint.is-docked-top-center .main-container{top:0}.is-xl-height-breakpoint .main-container{max-height:var(--arcgis-internal-popup-content-x-large-size)}.is-lg-height-breakpoint .main-container{max-height:var(--arcgis-internal-popup-content-large-size)}.is-sm-height-breakpoint .main-container,.is-xs-height-breakpoint .main-container{max-height:var(--arcgis-internal-popup-content-small-size)}.is-docked{flex-flow:row nowrap;inset:0;margin:var(--arcgis-layout-overlay-space-top, 0) var(--arcgis-layout-overlay-space-right, 0) var(--arcgis-layout-overlay-space-bottom, 0) var(--arcgis-layout-overlay-space-left, 0);padding:var(--calcite-spacing-lg) var(--calcite-spacing-lg) var(--calcite-spacing-xxxl) var(--calcite-spacing-lg);animation:docking-animation .25s ease-out}.is-docked-top-left,.is-docked-top-center,.is-docked-top-right,.is-docked-top-start,.is-docked-top-end{align-items:flex-start}.is-docked-bottom-left,.is-docked-bottom-center,.is-docked-bottom-right,.is-docked-bottom-start,.is-docked-bottom-end{align-items:flex-end}.is-docked-top-left,.is-docked-bottom-left,.is-docked-top-start,.is-docked-bottom-start{justify-content:flex-start}.is-docked-top-center,.is-docked-bottom-center{justify-content:center}.is-docked-top-right,.is-docked-bottom-right,.is-docked-top-end,.is-docked-bottom-end{justify-content:flex-end}.is-docked-top-left.is-rtl,.is-docked-top-right.is-rtl,.is-docked-bottom-left.is-rtl,.is-docked-bottom-right.is-rtl{flex-direction:row-reverse}.is-docked-top-left .main-container,.is-docked-top-right .main-container,.is-docked-bottom-left .main-container,.is-docked-bottom-right .main-container,.is-docked-top-start .main-container,.is-docked-top-end .main-container,.is-docked-bottom-start .main-container,.is-docked-bottom-end .main-container{max-height:80%}.is-docked-top-center .main-container,.is-docked-bottom-center .main-container{max-height:40%}@media(prefers-reduced-motion){.aligned-top-center,.aligned-top-left,.aligned-top-right,.aligned-bottom-center,.aligned-bottom-left,.aligned-bottom-right{animation:none}}@media(prefers-reduced-motion){.is-docked{animation:none}}@keyframes docking-animation{0%{opacity:0}25%{opacity:0}to{opacity:1}}@keyframes popup-intro-animation-down{0%{transform:translateY(calc(var(--calcite-spacing-xs) * -1));opacity:0}to{transform:translate(0);opacity:1}}@keyframes popup-intro-animation-up{0%{transform:translateY(var(--calcite-spacing-xs));opacity:0}to{transform:translate(0);opacity:1}}`;
function b(k) {
  return k && Number.parseInt(k.replaceAll(/[^-\d.]/gu, ""), 10) || 0;
}
const j = L(F), M = {
  buttonEnabled: !0,
  position: "auto",
  breakpoint: {
    width: 544
  }
};
class W extends S {
  constructor() {
    super(...arguments), this._direction = N(), this._featuresComponent = m(), this._dockAction = new B({ id: "popup-dock-action" }), this._mainContainerNode = m(), this._pointerNode = m(), this.messages = $(), this.viewModel = j(this), this._containerNode = m(), this._popupResizeObserver = new ResizeObserver(() => this._positionContainer()), this._positionContainer = () => {
      this.positionStyle = this._calculatePositionStyle();
    }, this.screenLocation = H(this), this.actions = this.viewModel.actions, this.active = this.viewModel.active, this.alignment = "auto", this.autoCloseEnabled = this.viewModel.autoCloseEnabled, this.autoDestroyDisabled = !1, this.collapsed = !1, this.content = this.viewModel.content, this.defaultPopupTemplateEnabled = this.viewModel.defaultPopupTemplateEnabled, this.dockEnabled = !1, this.featureCount = this.viewModel.featureCount, this.featureMenuOpen = this.viewModel.featureMenuOpen, this.featureMenuTitle = this.viewModel.featureMenuTitle, this.features = this.viewModel.features, this.goToOverride = this.viewModel.goToOverride, this.heading = this.viewModel.title, this.headingLevel = 2, this.hideActionBar = !1, this.hideCloseButton = !1, this.hideCollapseButton = !1, this.hideFeatureMenuHeading = !1, this.hideFeatureNavigation = !1, this.hideFeatureListLayerTitle = !1, this.hideHeading = !1, this.hideSpinner = !1, this.highlightDisabled = this.viewModel.highlightEnabled, this.includeDefaultActionsDisabled = this.viewModel.includeDefaultActions, this.initialDisplayMode = this.viewModel.initialDisplayMode, this.location = this.viewModel.location, this.open = this.viewModel.visible, this.promises = this.viewModel.promises, this.selectedFeature = this.viewModel.selectedFeature, this.selectedFeatureIndex = this.viewModel.selectedFeatureIndex, this.state = this.viewModel.state, this.topLayerDisabled = !1, this.updateLocationEnabled = this.viewModel.updateLocationEnabled, this.view = this.viewModel.view, this.arcgisPropertyChange = O()("active", "collapsed", "currentAlignment", "dockEnabled", "featureCount", "featureMenuOpen", "features", "promises", "selectedDrillInFeature", "selectedFeature", "selectedFeatureComponent", "selectedFeatureIndex", "state", "open"), this.arcgisReady = _(), this.arcgisClose = _(), this.arcgisTriggerAction = _();
  }
  static {
    this.properties = { screenLocation: 16, positionStyle: 16, _featureNavigationTop: 16, actions: 0, active: 0, alignment: 1, autoCloseEnabled: 5, autoDestroyDisabled: 5, collapsed: 5, content: 1, currentAlignment: 32, currentDockPosition: 32, defaultPopupTemplateEnabled: 5, dockEnabled: 5, dockOptions: 0, featureCount: 0, featureMenuOpen: 5, featureMenuTitle: 1, features: 0, goToOverride: 0, heading: 1, headingLevel: 9, hideActionBar: 5, hideCloseButton: 5, hideCollapseButton: 5, hideFeatureMenuHeading: 5, hideFeatureNavigation: 5, hideFeatureListLayerTitle: 5, hideHeading: 5, hideSpinner: 5, highlightDisabled: 5, includeDefaultActionsDisabled: 5, initialDisplayMode: 1, label: 1, location: 0, messageOverrides: 0, open: 5, promises: 0, referenceElement: 1, selectedDrillInFeature: 32, selectedFeature: 0, selectedFeatureComponent: 32, selectedFeatureIndex: 9, state: 32, topLayerDisabled: 5, updateLocationEnabled: 5, view: 0 };
  }
  static {
    this.styles = I;
  }
  get trackLocation() {
    return this.active && !this.dockEnabled;
  }
  get _featureNavigationTop() {
    const { currentAlignment: t, currentDockPosition: e } = this;
    return t === "bottom-left" || t === "bottom-center" || t === "bottom-right" || t === "bottom-start" || t === "bottom-end" || e === "top-left" || e === "top-center" || e === "top-right" || e === "top-start" || e === "top-end";
  }
  get currentAlignment() {
    return this._getCurrentAlignment();
  }
  get currentDockPosition() {
    return this._getCurrentDockPosition();
  }
  get dockOptions() {
    return this._dockOptions || M;
  }
  set dockOptions(t) {
    const e = { ...M }, i = this.viewModel?.view?.breakpoints, o = {};
    i && (o.width = i.xsmall, o.height = i.xsmall);
    const r = { ...e, ...t }, n = { ...e.breakpoint, ...o }, { breakpoint: s } = r;
    typeof s == "object" ? r.breakpoint = { ...n, ...s } : s && (r.breakpoint = n), this._dockOptions = r;
  }
  get selectedDrillInFeature() {
    return this._featuresComponent.value?.selectedDrillInFeature ?? null;
  }
  get selectedFeatureComponent() {
    return this._featuresComponent.value?.selectedFeatureComponent ?? null;
  }
  async handleViewClick(t) {
    return this.viewModel.handleViewClick(t);
  }
  async clear() {
    return this.viewModel.clear();
  }
  async destroy() {
    await this.manager.destroy(), this._dockAction.destroy(), this._popupResizeObserver.disconnect();
  }
  async fetchFeatures(t, e) {
    return await this._featuresComponent.value?.fetchFeatures(t, e);
  }
  async next() {
    return this.viewModel.next();
  }
  async previous() {
    return this.viewModel.previous();
  }
  async setFocus() {
    this._featuresComponent.value?.setFocus();
  }
  async triggerAction(t) {
    return await this._featuresComponent.value?.triggerAction(t);
  }
  loaded() {
    this.manager.onLifecycle(() => [
      l(() => [this.viewModel?.view?.widthBreakpoint, this.dockEnabled, this._direction], () => this._handleDockIcon(), { initial: !0 }),
      l(() => [this.dockEnabled, this.messages?.undock, this.messages?.dock], () => this._handleDockEnabled(), { initial: !0 }),
      l(() => [this.dockOptions, this._featuresComponent.value], () => {
        const { _dockAction: t, dockOptions: e } = this, i = this._featuresComponent.value?.headerActions;
        i?.remove(t), e.buttonEnabled && i?.add(t);
      }, { initial: !0 }),
      l(() => [
        this._pointerNode.value,
        this.alignment,
        this.currentAlignment,
        this.dockEnabled,
        this.dockOptions,
        this.screenLocation,
        this.viewModel?.active,
        this.viewModel?.location,
        this.viewModel?.view?.size
      ], () => this._positionContainer(), { initial: !0 }),
      l(() => this._mainContainerNode.value, (t) => {
        this._popupResizeObserver.disconnect(), t && this._popupResizeObserver.observe(t);
      }, { initial: !0 }),
      l(() => this.viewModel?.view?.size, (t, e) => this._updateDockEnabledForViewSize(t, e)),
      l(() => this.viewModel?.view?.ready, () => this._setDockEnabledForViewSize(this.dockOptions), { initial: !0 }),
      l(() => this.view?.map, (t) => {
        this.viewModel.map = t;
      }, { initial: !0 }),
      l(() => this.screenLocation, (t, e) => this._autoClose(t, e)),
      l(() => [this.active, this._containerNode.value, this.dockEnabled, this.topLayerDisabled], () => this._togglePopover(), { initial: !0 })
    ]);
  }
  _togglePopover() {
    this._containerNode.value?.togglePopover(this.active && !this.dockEnabled && !this.topLayerDisabled);
  }
  _handleClose(t) {
    t.stopPropagation(), this.arcgisClose.emit();
  }
  _autoClose(t, e) {
    this.autoCloseEnabled && t != null && e != null && (this.open = !1);
  }
  _handleArcgisPropertyChange(t) {
    t.detail.name === "collapsed" && (this.collapsed = t.target.collapsed);
  }
  _handleArcgisTriggerAction(t) {
    t.stopPropagation(), t.detail.action === this._dockAction && (this.dockEnabled = !this.dockEnabled), this.arcgisTriggerAction.emit({ action: t.detail.action });
  }
  _isOutsideView(t) {
    const { popupHeight: e, popupWidth: i, screenLocation: o, side: r, view: n } = t;
    return Number.isNaN(i) || Number.isNaN(e) || !n || !o ? !1 : r === "right" && o.x + i / 2 > n.width || r === "left" && o.x - i / 2 < 0 || r === "top" && o.y - e < 0 || r === "bottom" && o.y + e > n.height;
  }
  _getPointerSize() {
    const { _pointerNode: t } = this, e = t.value ? window.getComputedStyle(t.value) : null, i = b(e?.height), o = b(e?.width);
    return Math.hypot(i, o);
  }
  _calculateAutoAlignment(t) {
    if (t !== "auto")
      return t;
    const { _mainContainerNode: e, _popupRect: i, viewModel: o } = this, { view: r } = o, { screenLocation: n } = this;
    if (n == null || !r || !i || !e.value)
      return "top-center";
    const s = window.getComputedStyle(e.value), a = b(s.maxHeight), d = b(s.height), c = this._getPointerSize(), { height: p, width: v } = i, h = v + c, u = Math.max(p, a, d) + c, x = this._isOutsideView({
      popupHeight: u,
      popupWidth: h,
      screenLocation: n,
      side: "right",
      view: r
    }), z = this._isOutsideView({
      popupHeight: u,
      popupWidth: h,
      screenLocation: n,
      side: "left",
      view: r
    }), w = this._isOutsideView({
      popupHeight: u,
      popupWidth: h,
      screenLocation: n,
      side: "top",
      view: r
    }), D = this._isOutsideView({
      popupHeight: u,
      popupWidth: h,
      screenLocation: n,
      side: "bottom",
      view: r
    });
    return z ? w ? "bottom-right" : "top-right" : x ? w ? "bottom-left" : "top-left" : w ? D ? "top-center" : "bottom-center" : "top-center";
  }
  _callCurrentAlignment(t) {
    return typeof t == "function" ? t.call(this) : t;
  }
  _getCurrentAlignment() {
    return this.dockEnabled || !this.active || !this._mainContainerNode.value ? null : this._calculateAutoAlignment(this._callCurrentAlignment(this.alignment));
  }
  _callDockPosition(t) {
    return typeof t == "function" ? t.call(this) : t;
  }
  _getDockPosition() {
    return this._calculateAutoDockPosition(this._callDockPosition(this.dockOptions?.position));
  }
  _getCurrentDockPosition() {
    return this.dockEnabled && this.active ? this._getDockPosition() : null;
  }
  _calculateAutoDockPosition(t) {
    if (t !== "auto")
      return t;
    const e = this.viewModel?.view, i = "top-end";
    if (!e)
      return i;
    const { breakpoints: o } = e;
    return o?.xsmall && e.width <= o.xsmall ? "bottom-center" : i;
  }
  _getDockIcon() {
    const t = this._getDockPosition(), { _direction: e } = this;
    if (this.dockEnabled)
      return "minimize";
    switch (t) {
      case "top-left":
      case "bottom-left":
        return "dock-left";
      case "top-center":
        return "maximize";
      case "bottom-center":
        return "dock-bottom";
      case "top-right":
      case "bottom-right":
        return "dock-right";
      case "top-start":
      case "bottom-start":
        return e === "rtl" ? "dock-right" : "dock-left";
      case "top-end":
      case "bottom-end":
      default:
        return e === "rtl" ? "dock-left" : "dock-right";
    }
  }
  _handleDockIcon() {
    this._dockAction.icon = this._getDockIcon();
  }
  _handleDockEnabled() {
    this._dockAction.title = this.dockEnabled ? this.messages?.undock : this.messages?.dock;
  }
  _calculateAlignmentPosition(t, e, i, o) {
    if (t < 0 || e < 0 || !i || !o || t > i.width || e > i.height)
      return;
    const { _direction: r, currentAlignment: n } = this, s = this._getPointerSize(), a = s / 2;
    if (n === "top-left" || n === "top-start" && r === "ltr" || n === "top-end" && r === "rtl")
      return {
        top: e + i.top - (o.height + a),
        left: t + i.left - o.width - a
      };
    if (n === "top-center")
      return {
        top: e + i.top - (o.height + s),
        left: t + i.left - o.width / 2
      };
    if (n === "top-right" || n === "top-end" && r === "ltr" || n === "top-start" && r === "rtl")
      return {
        top: e + i.top - (o.height + a),
        left: t + i.left + a
      };
    if (n === "bottom-left" || n === "bottom-start" && r === "ltr" || n === "bottom-end" && r === "rtl")
      return {
        top: e + i.top + a,
        left: t + i.left - o.width - a
      };
    if (n === "bottom-center")
      return {
        top: e + i.top + s,
        left: t + i.left - o.width / 2
      };
    if (n === "bottom-right" || n === "bottom-end" && r === "ltr" || n === "bottom-start" && r === "rtl")
      return {
        top: e + i.top + a,
        left: t + i.left + a
      };
  }
  _getPositionStyleValue(t) {
    return t != null ? `${t}px` : "auto";
  }
  _calculatePositionStyle() {
    if (this.dockEnabled)
      return;
    const { screenLocation: t } = this;
    if (t == null)
      return;
    const e = this.view?.container?.getBoundingClientRect(), i = this._mainContainerNode.value?.getBoundingClientRect();
    this._popupRect = i;
    const o = this._calculateAlignmentPosition(t.x, t.y, e, i);
    if (o)
      return `inset: ${this._getPositionStyleValue(o.top)} ${this._getPositionStyleValue(o.right)} ${this._getPositionStyleValue(o.bottom)} ${this._getPositionStyleValue(o.left)};`;
  }
  _dockingThresholdCrossed(t, e, i) {
    const [o, r] = t, [n, s] = e, { width: a = 0, height: d = 0 } = i ?? {};
    return o <= a && n > a || o > a && n <= a || r <= d && s > d || r > d && s <= d;
  }
  _updateDockEnabledForViewSize(t, e) {
    if (!t || !e || !this.viewModel?.view?.ready)
      return;
    const { dockOptions: i } = this, o = i.breakpoint;
    typeof o == "object" && this._dockingThresholdCrossed(t, e, o) && this._setDockEnabledForViewSize(i);
  }
  _shouldDockAtCurrentViewSize(t) {
    const e = t.breakpoint, i = this.viewModel?.view?.ui;
    if (!i)
      return !1;
    const { width: o, height: r } = i;
    if (Number.isNaN(o) || Number.isNaN(r) || typeof e != "object")
      return !1;
    const n = typeof e.width == "number" && o <= (e.width ?? 0), s = typeof e.height == "number" && r <= (e.height ?? 0);
    return n || s;
  }
  _setDockEnabledForViewSize(t) {
    this.viewModel?.view?.ready && typeof t.breakpoint == "object" && (this.dockEnabled = this._shouldDockAtCurrentViewSize(t));
  }
  render() {
    const { _direction: t, dockEnabled: e, currentAlignment: i, currentDockPosition: o, positionStyle: r } = this, { active: n, view: s } = this.viewModel, a = n && e, d = n && !e, c = s?.widthBreakpoint, p = s?.heightBreakpoint, v = {
      "aligned-top-center": i === "top-center",
      "aligned-bottom-center": i === "bottom-center",
      "aligned-top-left": i === "top-left" || i === "top-start" && t === "ltr" || i === "top-end" && t === "rtl",
      "aligned-bottom-left": i === "bottom-left" || i === "bottom-start" && t === "ltr" || i === "bottom-end" && t === "rtl",
      "aligned-top-right": i === "top-right" || i === "top-end" && t === "ltr" || i === "top-start" && t === "rtl",
      "aligned-bottom-right": i === "bottom-right" || i === "bottom-end" && t === "ltr" || i === "bottom-start" && t === "rtl",
      "is-docked": a,
      shadow: d,
      "is-rtl": this._direction === "rtl",
      "is-docked-top-start": o === "top-start",
      "is-docked-top-end": o === "top-end",
      "is-docked-top-left": o === "top-left",
      "is-docked-top-center": o === "top-center",
      "is-docked-top-right": o === "top-right",
      "is-docked-bottom-left": o === "bottom-left",
      "is-docked-bottom-center": o === "bottom-center",
      "is-docked-bottom-right": o === "bottom-right",
      "is-docked-bottom-start": o === "bottom-start",
      "is-docked-bottom-end": o === "bottom-end",
      "is-xl-width-breakpoint": c === "xlarge",
      "is-lg-width-breakpoint": c === "large",
      "is-md-width-breakpoint": c === "medium",
      "is-sm-width-breakpoint": c === "small",
      "is-xs-width-breakpoint": c === "xsmall",
      "is-xl-height-breakpoint": p === "xlarge",
      "is-lg-height-breakpoint": p === "large",
      "is-md-height-breakpoint": p === "medium",
      "is-sm-height-breakpoint": p === "small",
      "is-xs-height-breakpoint": p === "xsmall"
    };
    return g`<div style=${T(r)} class=${y(C("root", v, {
      hidden: !n || !r && !e
    }))} .ariaHidden=${!n} .ariaLabel=${V(this.heading ?? "")} role=dialog popover=manual ${f(this._containerNode)}>${this._renderMainContainer()}${this._renderPointer()}</div>`;
  }
  _renderPointer() {
    return !this.dockEnabled || !this.active ? g`<div class="pointer" role=presentation><div class=${y(C("pointer-direction", "shadow"))} ${f(this._pointerNode)}></div></div>` : null;
  }
  _renderMainContainer() {
    const t = {
      shadow: this.dockEnabled
    };
    return g`<div class=${y(C("main-container", t))} ${f(this._mainContainerNode)}>${P(this.viewModel, g`<arcgis-features .collapsed=${this.collapsed} .featureNavigationTop=${this._featureNavigationTop} .headingLevel=${this.headingLevel} .hidden=${!this.active} .hideActionBar=${this.hideActionBar} .hideCloseButton=${this.hideCloseButton} .hideFeatureListLayerTitle=${this.hideFeatureListLayerTitle} .hideFeatureMenuHeading=${this.hideFeatureMenuHeading} .hideFeatureNavigation=${this.hideFeatureNavigation} .hideHeading=${this.hideHeading} .hideSpinner=${this.hideSpinner} @arcgisPropertyChange=${this._handleArcgisPropertyChange} @arcgisTriggerAction=${this._handleArcgisTriggerAction} @arcgisClose=${this._handleClose} .referenceElement=${this.referenceElement} .responsiveActionsEnabled=${!0} .showCollapseButton=${!this.hideCollapseButton} .viewModel=${this.viewModel} ${f(this._featuresComponent)}></arcgis-features>`)}</div>`;
  }
}
A("arcgis-popup", W);
export {
  W as ArcgisPopup
};
