// @arcgis/map-components/dist/components/arcgis-elevation-profile-settings/customElement.js
/* COPYRIGHT Esri - https://js.arcgis.com/5.0/LICENSE.txt */
import { c } from "../../chunks/runtime.js";
import { ref as o } from "lit/directives/ref.js";
import { css as r, html as t } from "lit";
import { LitElement as h, createEvent as p, nothing as s } from "@arcgis/lumina";
import { V as l } from "../../chunks/visible-element.js";
const g = r`:host{--arcgis-unit-select-inline-size: 180px}calcite-popover[calcite-hydrated-hidden]{inset-block-start:-999999px;inset-inline-start:-999999px}.popover-content{display:flex;flex-direction:column;gap:var(--calcite-spacing-md);padding:var(--calcite-spacing-md);max-width:300px}.uniform-chart-scaling-label{--calcite-label-margin-bottom: 0}.uniform-chart-scaling-checkbox-wrapper{width:var(--arcgis-unit-select-inline-size);flex-shrink:0}`;
class d extends h {
  constructor() {
    super(...arguments), this._onUniformChartScalingChange = (e) => {
      this.arcgisElevationProfileSettingsUniformChartScalingChange.emit(e.currentTarget.checked);
    }, this._onActionRef = (e) => {
      this._actionElement = e ?? void 0;
    }, this._popoverIsOpen = !1, this.hideDistanceDisplayUnitSelect = !1, this.hideElevationDisplayUnitSelect = !1, this.hideUniformChartScalingToggle = !1, this.arcgisElevationProfileSettingsUniformChartScalingChange = p();
  }
  static {
    this.properties = { _popoverIsOpen: 16, _actionElement: 16, viewModel: 0, hideDistanceDisplayUnitSelect: 5, hideElevationDisplayUnitSelect: 5, hideUniformChartScalingToggle: 5, messages: 0, uniformChartScaling: 5, unitOptionsDistance: 0, unitOptionsElevation: 0 };
  }
  static {
    this.styles = g;
  }
  async destroy() {
    await this.manager.destroy();
  }
  _onPopoverClose() {
    this._popoverIsOpen = !1;
  }
  _onPopoverOpen() {
    this._popoverIsOpen = !0;
  }
  render() {
    const { messages: e } = this;
    if (!e || this.hideDistanceDisplayUnitSelect && this.hideElevationDisplayUnitSelect && this.hideUniformChartScalingToggle)
      return;
    const i = (this._popoverIsOpen ? e.hideSettings : e.showSettings) ?? "";
    return t`<calcite-action data-testid=settings-button icon=gear .text=${i} title=${i ?? s} ${o(this._onActionRef)}></calcite-action>${this._actionElement ? t`<calcite-popover auto-close .label=${i} placement=bottom-end .referenceElement=${this._actionElement} @calcitePopoverOpen=${this._onPopoverOpen} @calcitePopoverClose=${this._onPopoverClose}><div class="popover-content">${this._renderUnitSelector()}${this._renderUniformChartScalingToggle()}</div></calcite-popover>` : void 0}`;
  }
  _renderUnitSelector() {
    const { viewModel: e, messages: i } = this;
    if (!e || this.hideDistanceDisplayUnitSelect && this.hideElevationDisplayUnitSelect)
      return;
    const n = "inline-space-between";
    return t`${l({ hidden: this.hideDistanceDisplayUnitSelect, children: t`<arcgis-unit-select .options=${this.unitOptionsDistance} .selectLabel=${i.unitDistanceSelectLabel} .value=${e.distanceUnit} .layout=${n} @arcgisUnitSelectChange=${(a) => {
      e.distanceUnit = a.detail;
    }}></arcgis-unit-select>` })}${l({ hidden: this.hideElevationDisplayUnitSelect, children: t`<arcgis-unit-select .options=${this.unitOptionsElevation} .selectLabel=${i.unitElevationSelectLabel} .value=${e.elevationUnit} .layout=${n} @arcgisUnitSelectChange=${(a) => {
      e.elevationUnit = a.detail;
    }}></arcgis-unit-select>` })}`;
  }
  _renderUniformChartScalingToggle() {
    const { messages: e, hideUniformChartScalingToggle: i, uniformChartScaling: n } = this;
    if (i)
      return;
    const a = e?.[n ? "uniformChartScalingDisable" : "uniformChartScalingEnable"] ?? "";
    return t`<calcite-label class="uniform-chart-scaling-label" layout=inline-space-between>${e?.uniformChartScalingLabel}<div class="uniform-chart-scaling-checkbox-wrapper"><calcite-checkbox .checked=${n} title=${a ?? s} @calciteCheckboxChange=${this._onUniformChartScalingChange}></calcite-checkbox></div></calcite-label>`;
  }
}
c("arcgis-elevation-profile-settings", d);
export {
  d as ArcgisElevationProfileSettings
};
