// @arcgis/map-components/dist/components/arcgis-shadow-cast-total-duration-config/customElement.js
/* COPYRIGHT Esri - https://js.arcgis.com/5.0/LICENSE.txt */
import { c as a } from "../../chunks/runtime.js";
import { css as r, html as o } from "lit";
import { LitElement as n } from "@arcgis/lumina";
import { V as s } from "../../chunks/visible-element.js";
const d = r`:host{--calcite-label-margin-bottom: 0;gap:var(--calcite-spacing-md);flex-direction:column;display:flex}calcite-label{display:block}`;
class m extends n {
  constructor() {
    super(...arguments), this.scale = "s", this.hideColorPicker = !1, this.hideModeSelect = !1;
  }
  static {
    this.properties = { options: 0, scale: 1, hideColorPicker: 5, hideModeSelect: 5, messages: 0 };
  }
  static {
    this.styles = d;
  }
  render() {
    return o`${s({ hidden: this.hideModeSelect, children: this._renderModeSelect() })}${s({ hidden: this.hideColorPicker, children: o`<arcgis-color-select .color=${this.options.color} .label=${this.messages?.totalDuration?.colorLabel} @arcgisColorChanged=${(e) => {
      this.options.color = e.detail;
    }}></arcgis-color-select>` })}`;
  }
  _renderModeSelect() {
    const { options: e, scale: l } = this, { mode: c } = e, t = this.messages?.totalDuration;
    return o`<calcite-label .scale=${l}>${t?.modeLabel}<calcite-segmented-control .scale=${l} width=full @calciteSegmentedControlChange=${(i) => {
      e.mode = i.currentTarget.value;
    }}><calcite-segmented-control-item value=hourly .checked=${c === "hourly"}>${t?.hourlyLabel}</calcite-segmented-control-item><calcite-segmented-control-item value=continuous .checked=${c === "continuous"}>${t?.continuousLabel}</calcite-segmented-control-item></calcite-segmented-control></calcite-label>`;
  }
}
a("arcgis-shadow-cast-total-duration-config", m);
export {
  m as ArcgisShadowCastTotalDurationConfig
};
