// @arcgis/map-components/dist/components/arcgis-weather/customElement.js
/* COPYRIGHT Esri - https://js.arcgis.com/5.0/LICENSE.txt */
import { c as T } from "../../chunks/runtime.js";
import { css as A, html as c } from "lit";
import { ref as O } from "lit/directives/ref.js";
import { useDirection as S, usePropertyChange as B } from "@arcgis/lumina/controllers";
import { classes as L } from "@arcgis/toolkit/dom";
import { LitElement as E, createEvent as F, nothing as v, safeClassMap as b, safeStyleMap as D } from "@arcgis/lumina";
import { u as j } from "../../chunks/useT9n.js";
import { m as R } from "../../chunks/useViewModel.js";
import { H } from "../../chunks/heading.js";
import { g as x } from "../../chunks/globalCss.js";
import P from "@arcgis/core/core/Accessor.js";
import { property as p, subclass as V } from "@arcgis/core/core/accessorSupport/decorators.js";
import { watch as I } from "@arcgis/core/core/reactiveUtils.js";
import C from "@arcgis/core/views/3d/environment/CloudyWeather.js";
import $ from "@arcgis/core/views/3d/environment/FoggyWeather.js";
import k from "@arcgis/core/views/3d/environment/RainyWeather.js";
import M from "@arcgis/core/views/3d/environment/SnowyWeather.js";
import z from "@arcgis/core/views/3d/environment/SunnyWeather.js";
const K = A`.esri-widget__heading{color:var(--calcite-color-text-2);margin:0 0 .5rem;font-weight:var(--calcite-font-weight-semibold)}h1.esri-widget__heading{font-size:var(--calcite-font-size-xl)}h2.esri-widget__heading{font-size:var(--calcite-font-size-md)}h3.esri-widget__heading,h4.esri-widget__heading,h5.esri-widget__heading,h6.esri-widget__heading{font-size:var(--calcite-font-size)}.esri-widget{box-sizing:border-box;color:var(--calcite-color-text-2);background-color:var(--calcite-color-foreground-1);font-size:var(--calcite-font-size);line-height:var(--calcite-font-line-height-relative-tight)}.esri-widget *,.esri-widget :before,.esri-widget :after{box-sizing:inherit}.esri-widget.esri-widget--disabled>*{opacity:.4;pointer-events:none}.esri-widget.esri-widget--compact{--esri-widget-padding-y: 0;--esri-widget-padding-x: 0}.esri-widget__anchor{color:var(--calcite-color-text-2)}.esri-widget__anchor:hover{color:var(--calcite-color-text-3)}.esri-widget__anchor:hover,.esri-widget__anchor:focus{text-decoration:underline}.esri-widget__anchor--disabled{opacity:.4;pointer-events:none}.esri-widget__header{background-color:var(--calcite-color-foreground-1);flex:1 0 100%;justify-content:space-between;align-items:center;width:100%;min-height:48px;padding:12px 0 12px 15px;display:flex}.esri-widget__heading{word-break:break-word}.esri-widget__header .esri-widget__heading{width:calc(100% - 48px);text-align:initial;color:var(--calcite-color-text-1);align-items:center;margin:0;padding:0;display:flex}.esri-widget__header-button{appearance:none;cursor:pointer;color:var(--calcite-color-text-2);background-color:var(--calcite-color-foreground-1);border:none;flex:0 0 48px;justify-content:center;align-self:stretch;align-items:center;width:48px;padding:0;display:flex}.esri-widget__footer{background-color:var(--calcite-color-foreground-1);justify-content:center;align-items:center;width:100%;min-height:48px;display:flex}.esri-widget__footer-pagination{color:var(--calcite-color-text-2);justify-content:space-between;align-items:center;display:flex}.esri-widget__footer-pagination-previous-button,.esri-widget__footer-pagination-next-button{cursor:pointer;background-color:var(--calcite-color-foreground-1);border:0;margin:0 7px;padding:12px 15px;transition:background-color .125s ease-in-out}.esri-widget__footer-pagination-previous-button:hover,.esri-widget__footer-pagination-previous-button:focus,.esri-widget__footer-pagination-next-button:hover,.esri-widget__footer-pagination-next-button:focus{color:var(--calcite-color-text-1);background-color:var(--calcite-color-foreground-1)}.esri-menu{visibility:hidden;z-index:1;background-clip:padding;background-color:var(--calcite-color-foreground-1);background-clip:padding-box;width:100%;max-height:0;margin:2px 0 0;padding:0;font-size:var(--calcite-font-size);line-height:var(--calcite-spacing-xl);position:absolute;top:100%;left:0;overflow:hidden;box-shadow:0 1px 2px #0000004d}.esri-menu .esri-menu__header{color:var(--calcite-color-text-inverse);background-color:var(--calcite-color-inverse);padding:6px 12px}.esri-menu__list{margin:0;padding:0;list-style:none}.esri-menu__list-item{cursor:pointer;border-top:1px solid #adadad4d;border-left:3px solid #0000;padding:.8em 1em}.esri-menu__list-item [class^=esri-icon-],.esri-menu__list-item [class*=" esri-icon-"]{padding-right:2.8px}.esri-menu__list-item:first-child{border-top:none}.esri-menu__list-item--focus,.esri-menu__list-item:hover,.esri-menu__list-item:focus{background-color:var(--calcite-color-foreground-1)}.esri-menu__list-item:active{background-color:var(--calcite-color-foreground-current)}.esri-menu__list-item--active,.esri-menu__list-item--active:hover,.esri-menu__list-item--active:focus{background-color:var(--calcite-color-foreground-1);border-left-color:var(--calcite-color-brand)}.esri-widget__table{border-collapse:collapse;color:var(--calcite-color-text-1);border:none;width:100%;line-height:var(--calcite-font-line-height-relative-tight)}.esri-widget__table tr:nth-child(odd){background-color:var(--calcite-color-transparent)}.esri-widget__table tr:nth-child(2n){background-color:var(--calcite-color-transparent-press)}.esri-widget__table tr a{color:var(--calcite-color-text-2)}.esri-widget__table tr a:hover,.esri-widget__table tr a:focus{color:var(--calcite-color-text-1)}.esri-widget__table tr td,.esri-widget__table tr th{vertical-align:top;word-break:break-word;padding:.5em .7em;font-size:var(--calcite-font-size-sm);font-weight:var(--calcite-font-weight-regular)}.esri-widget__table tr th{text-align:left;border-right:3px solid #0000000d;width:50%}.esri-widget__table tr td{width:50%}.esri-input{color:var(--calcite-color-text-1);background-color:var(--calcite-color-foreground-1);border:1px solid #959595;font-size:var(--calcite-font-size)}.esri-input::-ms-clear{display:none}.esri-input::placeholder{opacity:1;color:var(--calcite-color-text-2)}.esri-input:placeholder-shown{text-overflow:ellipsis}.esri-input[type=text],.esri-input[type=password],.esri-input[type=number]{height:32px;padding:0 .5em}calcite-combobox.esri-input{border:none}.esri-widget__content--empty{text-align:center;color:var(--calcite-color-border-input);flex-flow:column wrap;align-items:center;padding:18px 22px;display:flex}.esri-widget__content--empty h1.esri-widget__heading,.esri-widget__content--empty h2.esri-widget__heading,.esri-widget__content--empty h3.esri-widget__heading,.esri-widget__content--empty h4.esri-widget__heading,.esri-widget__content--empty h5.esri-widget__heading{font-weight:var(--calcite-font-weight-regular)}.esri-widget__content-illustration--empty{width:128px;padding:1rem 0}.esri-widget__content-icon--empty{width:32px;padding:.5rem 0}.esri-select{appearance:none;cursor:pointer;color:var(--calcite-color-text-2);background:var(--calcite-color-foreground-1) url("data:image/svg+xml,%3Csvg xmlns='http://www.w3.org/2000/svg' viewBox='-8 -8 32 32' width='32' height='32'%3E%3Cpath d='M8 11.207l-4-4V5.793l4 4 4-4v1.414z' fill='%23adadad' fillrule='nonzero'/%3E%3C/svg%3E") 100% no-repeat;border:1px solid #959595;border-radius:0;width:100%;height:32px;margin:0;padding:0 .5em;font-family:inherit;font-size:.85em;display:block}.esri-select[disabled]{cursor:auto}.esri-select::-ms-expand{display:none}.esri-disabled a,.esri-disabled [role=menu],.esri-disabled [role=checkbox],.esri-disabled [class^=esri-icon-],.esri-disabled [class*=" esri-icon-"]{color:var(--calcite-color-text-3)}.keynav-mode .esri-widget:focus,.keynav-mode .esri-widget :focus{z-index:1;outline-offset:2px;outline:2px solid #adadad}.esri-widget__loader-animation{border:.15em solid #adadad4d;border-top-color:var(--calcite-color-border-input);border-radius:100%;width:1em;height:1em;font-size:var(--calcite-font-size-md);animation:1.25s linear infinite esri-rotate;display:block;transform:translateZ(0)}:dir(rtl) .esri-widget__table th{text-align:right;border-left:3px solid #0000000d;border-right:none}:dir(rtl) .esri-select{background-position:0}:dir(rtl) .esri-menu__list-item{border-left:none;border-right:3px solid #0000}:dir(rtl) .esri-menu__list-item [class^=esri-icon-],:dir(rtl) .esri-menu__list-item [class*=" esri-icon-"]{padding-left:2px;padding-right:0}:dir(rtl) .esri-menu__list-item--active{border-right-color:var(--calcite-color-brand)}.esri-icon-font-fallback-text{clip:rect(0 0 0 0);width:1px;height:1px;position:absolute;overflow:hidden}.esri-rotating{animation:1.25s linear infinite esri-rotate;transform:translateZ(0)}.esri-clearfix:before,.esri-clearfix:after{content:" ";display:table}.esri-clearfix:after{clear:both}.esri-interactive{cursor:pointer}.esri-hidden{display:none!important}.esri-invisible{visibility:hidden!important}.esri-offscreen{position:absolute;top:-999em;left:-999em}.esri-match-height{flex-direction:column;flex:auto;display:flex;overflow:hidden}:host{width:300px;min-width:180px}.root{position:relative;padding:var(--arcgis-internal-padding);container-type:inline-size;width:100%}.content{display:flex;flex-direction:column;gap:var(--calcite-spacing-lg)}.content.error{visibility:hidden;height:0;overflow:hidden;pointer-events:none}.weather-type-selector{--button-gap: var(--calcite-spacing-sm);display:flex;flex-wrap:wrap;justify-content:center;width:100%;margin-inline:auto;gap:var(--button-gap);>calcite-button{display:block;position:relative;height:var(--button-size);flex-shrink:0;flex-grow:0;--total-gap: calc((var(--weather-type-count) - 1) * var(--button-gap));--basis: clamp(24px, calc((100% - var(--total-gap)) / var(--weather-type-count)), 64px);flex-basis:round(down,var(--basis),1px)}}@container (max-width: 280px){.weather-type-selector{--button-gap: var(--calcite-spacing-xs)}}@container (max-width: 220px){.weather-type-selector{--button-gap: var(--calcite-spacing-xxs)}}.options{display:flex;flex-direction:column;gap:var(--calcite-spacing-md);calcite-label{--calcite-label-margin-bottom: 0}&.hidden{display:none}}.warning-title{position:absolute;width:1px;height:1px;padding:0;margin:-1px;overflow:hidden;clip:rect(0,0,0,0);border:0}`;
var q = Object.defineProperty, Z = Object.getOwnPropertyDescriptor, d = (o, e, t, i) => {
  for (var r = i > 1 ? void 0 : i ? Z(e, t) : e, n = o.length - 1, a; n >= 0; n--)
    (a = o[n]) && (r = (i ? a(e, t, r) : a(r)) || r);
  return i && r && q(e, t, r), r;
};
const U = {
  sunny: z,
  cloudy: C,
  rainy: k,
  snowy: M,
  foggy: $
}, u = Object.keys(U);
let s = class extends P {
  //#region Lifecycle
  constructor(o) {
    super(o), this._settingWeather = !1, this._weatherByType = {
      sunny: new z(),
      cloudy: new C(),
      rainy: new k(),
      snowy: new M(),
      foggy: new $()
    };
  }
  initialize() {
    this.addHandles([
      // When the weather is changed directly in the environment (not in the VM),
      // we need to update the corresponding weather instance in the view model.
      I(
        () => this.current,
        (o) => {
          this._settingWeather || o == null || (this._weatherByType = {
            ...this._weatherByType,
            [o.type]: o
          });
        },
        { sync: !0, initial: !0 }
      )
    ]);
  }
  get state() {
    if (!this.view?.ready)
      return "disabled";
    if (this.error != null)
      switch (this.error) {
        // When the weather is not visible because of camera position, we don't
        // consider it to be a fatal error so we still want to display the full
        // UI, just with a warning at the bottom.
        case "not-visible":
          return "ready";
        // Other errors make the widget unusable, so we consider them to be fatal
        // and show only the error message.
        default:
          return "error";
      }
    return "ready";
  }
  get disabled() {
    return this.state === "disabled";
  }
  get weatherByType() {
    return this._weatherByType;
  }
  get error() {
    const o = this.view;
    return o?.ready ? o.type !== "3d" ? "unsupported-view" : o.viewingMode === "local" ? "local-scene" : o.environment.atmosphereEnabled ? o.environment.weatherAvailable ? null : "not-visible" : "no-atmosphere" : null;
  }
  get current() {
    return this._environment?.weather ?? this._weatherByType.sunny;
  }
  set current(o) {
    const e = this._environment;
    e && (e.weather = o);
  }
  get _environment() {
    const o = this.view;
    return o?.type === "3d" ? o.environment : null;
  }
  //#endregion
  //#region Public Methods
  setWeatherByType(o) {
    const e = this._environment;
    e != null && (this._settingWeather = !0, e.weather = this._weatherByType[o], this._settingWeather = !1);
  }
  setSnowCoverEnabled(o) {
    const e = this.current;
    "snowCover" in e && (e.snowCover = o ? "enabled" : "disabled");
  }
  //#endregion
};
d([
  p()
], s.prototype, "view", 2);
d([
  p()
], s.prototype, "state", 1);
d([
  p()
], s.prototype, "disabled", 1);
d([
  p()
], s.prototype, "weatherByType", 1);
d([
  p()
], s.prototype, "error", 1);
d([
  p()
], s.prototype, "current", 1);
d([
  p()
], s.prototype, "_weatherByType", 2);
d([
  p()
], s.prototype, "_environment", 1);
s = d([
  V("components.weather.WeatherViewModel")
], s);
function g({ label: o, ariaLabel: e, value: t, disabled: i, onChange: r }) {
  const n = (l) => {
    const h = l.currentTarget.value;
    r?.(h ?? 0);
  };
  return c`<calcite-label .scale=${"m"}><calcite-slider .disabled=${i} .ariaLabel=${e} .labelText=${o} .value=${t} min=0 max=1 step=0.01 .scale=${"m"} @calciteSliderChange=${n} @calciteSliderInput=${n}></calcite-slider></calcite-label>`;
}
const G = R(s), J = {
  sunny: "brightness",
  cloudy: "partly-cloudy",
  rainy: "rain",
  snowy: "snow",
  foggy: "fog"
};
class N extends E {
  constructor() {
    super(...arguments), this.viewModel = G(this), this.messages = j(), this.direction = S(), this._resizeAnimationFrame = void 0, this._resizeObserver = new ResizeObserver((e) => {
      for (const t of e)
        t.target instanceof HTMLElement && this._updateButtonMetrics(t.contentRect.width);
    }), this._buttonWidth = 0, this.view = this.viewModel.view, this.autoDestroyDisabled = !1, this.headingLevel = 4, this.hideHeader = !1, this.icon = "partly-cloudy", this.label = this.messages.componentLabel, this.state = this.viewModel.state, this.arcgisPropertyChange = B()("state"), this.arcgisReady = F();
  }
  static {
    this.properties = { _buttonWidth: 16, view: 0, autoDestroyDisabled: 5, headingLevel: 9, hideHeader: 5, icon: 1, label: 1, messageOverrides: 0, referenceElement: 1, state: 32 };
  }
  static {
    this.styles = K;
  }
  async destroy() {
    await this.manager.destroy();
  }
  _makeWeatherTypeButtonRef() {
    let e;
    return (t) => {
      t ? (e = t, this._resizeObserver.observe(t), this._updateButtonMetrics(t.clientWidth)) : e && (this._resizeObserver.unobserve(e), e = void 0);
    };
  }
  _updateButtonMetrics(e) {
    this._resizeAnimationFrame && cancelAnimationFrame(this._resizeAnimationFrame), this._resizeAnimationFrame = requestAnimationFrame(() => {
      this._buttonWidth = e, this._resizeAnimationFrame = void 0;
    });
  }
  _getWarningMessage() {
    switch (this.viewModel.error) {
      case "not-visible": {
        const { view: e } = this.viewModel;
        return e?.type === "3d" && e.environment.atmosphereEnabled ? this.messages.error?.notVisible : void 0;
      }
      default:
        return;
    }
  }
  _getErrorMessage() {
    const e = this.messages.error;
    if (e)
      switch (this.viewModel.error) {
        case "local-scene":
          return e.localScene;
        case "no-atmosphere":
          return e.noAtmosphere;
        case "unsupported-view":
          return e.unsupported;
        default:
          return;
      }
  }
  render() {
    return c`<div aria-label=${this.messages.componentLabelAccessible ?? v} class=${b({
      root: !0,
      [x.widget]: !0,
      [x.widgetDisabled]: this.viewModel.disabled
    })} role=region>${this._renderContent()}</div>`;
  }
  _renderContent() {
    const { headingLevel: e, messages: t, viewModel: i } = this, r = this._getWarningMessage(), n = i.current, a = i.state === "error";
    return c`${this.hideHeader ? void 0 : H({ level: e, children: t.componentLabel })}${a ? c`<div>${this._getErrorMessage()}</div>` : void 0}<div class=${b(L({ content: !0, error: a }))}><div .ariaLabel=${t.weatherType} class="weather-type-selector" role=radiogroup style=${D({
      "--weather-type-count": String(u.length),
      "--button-size": `${this._buttonWidth}px`
    })}>${u.map(this._renderWeatherTypeButton, this)}</div>${Object.values(i.weatherByType).map((l) => c`<div class=${b({
      options: !0,
      hidden: l.type !== n?.type
    })}>${this._renderOptions(l)}</div>`)}${r ? this._renderWarning(r) : void 0}</div>`;
  }
  _renderWeatherTypeButton(e) {
    const t = this.messages[e]?.label ?? "", { disabled: i } = this.viewModel, r = !i && this.viewModel.current.type === e;
    return c`<calcite-button role=radio .label=${t} .name=${t} .appearance=${r ? "solid" : "outline"} .ariaChecked=${r ? "true" : "false"} tabindex=${(r ? void 0 : -1) ?? v} .disabled=${i} .iconStart=${J[e]} .kind=${r ? "brand" : "neutral"} width=full .scale=${this._buttonWidth < 32 ? "s" : this._buttonWidth < 64 ? "m" : "l"} @keydown=${(n) => {
      const a = n.key;
      let l = ["ArrowRight", "ArrowDown"], h = ["ArrowLeft", "ArrowUp"];
      if (this.direction === "rtl") {
        const m = l;
        l = h, h = m;
      }
      const W = l.includes(a), w = h.includes(a);
      if (W || w) {
        n.preventDefault();
        const m = u.indexOf(e), f = u.length, _ = (m + (w ? -1 : 1) + f) % f;
        this.viewModel.setWeatherByType(u[_]);
        const y = n.currentTarget.parentElement?.parentElement;
        y && Array.from(y.querySelectorAll("calcite-button") ?? [])[_]?.setFocus();
      }
    }} @click=${(n) => {
      if (r)
        return;
      this.viewModel.setWeatherByType(e);
      const a = n.currentTarget;
      requestAnimationFrame(() => {
        this.viewModel.current.type === e && a.setFocus();
      });
    }} title=${t ?? v} ${O(this._makeWeatherTypeButtonRef())}></calcite-button>`;
  }
  _renderOptions(e) {
    switch (e.type) {
      case "sunny":
        return this._renderSunnyOptions(e);
      case "cloudy":
        return this._renderCloudyOptions(e);
      case "rainy":
        return this._renderRainyOptions(e);
      case "snowy":
        return this._renderSnowyOptions(e);
      case "foggy":
        return this._renderFoggyOptions(e);
    }
  }
  _renderSunnyOptions(e) {
    const t = this.messages.sunny;
    return g({ disabled: this.viewModel.disabled, ariaLabel: t?.cloudCoverAria, label: t?.cloudCover, value: e.cloudCover, onChange: (i) => {
      e.cloudCover = i;
    } });
  }
  _renderCloudyOptions(e) {
    const t = this.messages.cloudy;
    return g({ disabled: this.viewModel.disabled, ariaLabel: t?.cloudCoverAria, label: t?.cloudCover, value: e.cloudCover, onChange: (i) => {
      e.cloudCover = i;
    } });
  }
  _renderRainyOptions(e) {
    const { disabled: t } = this.viewModel, i = this.messages.rainy;
    return c`${g({ disabled: t, ariaLabel: i?.cloudCoverAria, label: i?.cloudCover, value: e.cloudCover, onChange: (r) => {
      e.cloudCover = r;
    } })}${g({ disabled: t, ariaLabel: i?.precipitationAria, label: i?.precipitation, value: e.precipitation, onChange: (r) => {
      e.precipitation = r;
    } })}`;
  }
  _renderSnowyOptions(e) {
    const { disabled: t } = this.viewModel, i = this.messages.snowy;
    return c`${g({ disabled: t, ariaLabel: i?.cloudCoverAria, label: i?.cloudCover, value: e.cloudCover, onChange: (r) => {
      e.cloudCover = r;
    } })}${g({ disabled: t, ariaLabel: i?.precipitationAria, label: i?.precipitation, value: e.precipitation, onChange: (r) => {
      e.precipitation = r;
    } })}<calcite-label layout=inline title=${i?.snowCoverTooltip ?? v}><calcite-checkbox .checked=${e.snowCover === "enabled"} .disabled=${t} @calciteCheckboxChange=${(r) => {
      this.viewModel.setSnowCoverEnabled(r.currentTarget.checked);
    }}></calcite-checkbox>${i?.snowCover}</calcite-label>`;
  }
  _renderFoggyOptions(e) {
    const t = this.messages.foggy;
    return g({ disabled: this.viewModel.disabled, ariaLabel: t?.fogStrengthAria, label: t?.fogStrength, value: e.fogStrength, onChange: (i) => {
      e.fogStrength = i;
    } });
  }
  _renderWarning(e) {
    return c`<calcite-notice aria-live=polite open icon=exclamation-mark-triangle-f kind=warning><div slot=title class="warning-title">${this.messages.warningTitle}</div><div slot=message>${e}</div></calcite-notice>`;
  }
}
T("arcgis-weather", N);
export {
  N as ArcgisWeather
};
