// @arcgis/map-components-react/dist/index.js
import i from "react";
import { createComponent as a } from "@lit/react";
import { makeReactWrapperFactory as g, getReactWrapperOptions as r } from "@arcgis/lumina";
const e = /* @__PURE__ */ g(i, a), n = /* @__PURE__ */ e(
  r(
    "arcgis-area-measurement-2d",
    {
      onArcgisPropertyChange: "arcgisPropertyChange",
      onArcgisReady: "arcgisReady"
    }
  )
), t = /* @__PURE__ */ e(
  r(
    "arcgis-area-measurement-3d",
    {
      onArcgisPropertyChange: "arcgisPropertyChange",
      onArcgisReady: "arcgisReady"
    }
  )
), A = /* @__PURE__ */ e(
  r(
    "arcgis-attachments",
    {
      onArcgisPropertyChange: "arcgisPropertyChange",
      onArcgisReady: "arcgisReady"
    }
  )
), y = /* @__PURE__ */ e(
  r(
    "arcgis-basemap-gallery",
    {
      onArcgisPropertyChange: "arcgisPropertyChange",
      onArcgisReady: "arcgisReady"
    }
  )
), l = /* @__PURE__ */ e(
  r(
    "arcgis-basemap-layer-list",
    {
      onArcgisClose: "arcgisClose",
      onArcgisPropertyChange: "arcgisPropertyChange",
      onArcgisReady: "arcgisReady",
      onArcgisTriggerAction: "arcgisTriggerAction"
    }
  )
), d = /* @__PURE__ */ e(
  r(
    "arcgis-basemap-toggle",
    {
      onArcgisPropertyChange: "arcgisPropertyChange",
      onArcgisReady: "arcgisReady"
    }
  )
), C = /* @__PURE__ */ e(
  r(
    "arcgis-bookmarks",
    {
      onArcgisBookmarkEdit: "arcgisBookmarkEdit",
      onArcgisBookmarkSelect: "arcgisBookmarkSelect",
      onArcgisClose: "arcgisClose",
      onArcgisPropertyChange: "arcgisPropertyChange",
      onArcgisReady: "arcgisReady"
    }
  )
), h = /* @__PURE__ */ e(
  r(
    "arcgis-building-explorer",
    {
      onArcgisBuildingLevelChange: "arcgisBuildingLevelChange",
      onArcgisBuildingPhaseChange: "arcgisBuildingPhaseChange",
      onArcgisReady: "arcgisReady"
    }
  )
), p = /* @__PURE__ */ e(
  r(
    "arcgis-catalog-layer-list",
    {
      onArcgisClose: "arcgisClose",
      onArcgisPropertyChange: "arcgisPropertyChange",
      onArcgisReady: "arcgisReady",
      onArcgisTriggerAction: "arcgisTriggerAction"
    }
  )
), w = /* @__PURE__ */ e(
  r(
    "arcgis-compass",
    {
      onArcgisPropertyChange: "arcgisPropertyChange",
      onArcgisReady: "arcgisReady"
    }
  )
), R = /* @__PURE__ */ e(
  r(
    "arcgis-coordinate-conversion",
    {
      onArcgisConversionChange: "arcgisConversionChange",
      onArcgisPropertyChange: "arcgisPropertyChange",
      onArcgisReady: "arcgisReady"
    }
  )
), V = /* @__PURE__ */ e(
  r(
    "arcgis-daylight",
    {
      onArcgisReady: "arcgisReady",
      onArcgisUserDateTimeChange: "arcgisUserDateTimeChange"
    }
  )
), P = /* @__PURE__ */ e(
  r(
    "arcgis-directional-pad",
    {
      onArcgisPropertyChange: "arcgisPropertyChange",
      onArcgisReady: "arcgisReady"
    }
  )
), u = /* @__PURE__ */ e(
  r(
    "arcgis-directions",
    {
      onArcgisPropertyChange: "arcgisPropertyChange",
      onArcgisReady: "arcgisReady"
    }
  )
), m = /* @__PURE__ */ e(
  r(
    "arcgis-direct-line-measurement-3d",
    {
      onArcgisPropertyChange: "arcgisPropertyChange",
      onArcgisReady: "arcgisReady"
    }
  )
), T = /* @__PURE__ */ e(
  r(
    "arcgis-distance-measurement-2d",
    {
      onArcgisPropertyChange: "arcgisPropertyChange",
      onArcgisReady: "arcgisReady"
    }
  )
), S = /* @__PURE__ */ e(
  r(
    "arcgis-editor",
    {
      onArcgisPropertyChange: "arcgisPropertyChange",
      onArcgisReady: "arcgisReady",
      onArcgisSketchCreate: "arcgisSketchCreate",
      onArcgisSketchUpdate: "arcgisSketchUpdate"
    }
  )
), D = /* @__PURE__ */ e(
  r(
    "arcgis-elevation-profile",
    {
      onArcgisPropertyChange: "arcgisPropertyChange",
      onArcgisReady: "arcgisReady"
    }
  )
), b = /* @__PURE__ */ e(
  r(
    "arcgis-expand",
    {
      onArcgisPropertyChange: "arcgisPropertyChange",
      onArcgisReady: "arcgisReady"
    }
  )
), L = /* @__PURE__ */ e(
  r(
    "arcgis-feature",
    {
      onArcgisPropertyChange: "arcgisPropertyChange",
      onArcgisReady: "arcgisReady"
    }
  )
), k = /* @__PURE__ */ e(
  r(
    "arcgis-feature-form",
    {
      onArcgisPropertyChange: "arcgisPropertyChange",
      onArcgisReady: "arcgisReady",
      onArcgisSubmit: "arcgisSubmit",
      onArcgisValueChange: "arcgisValueChange"
    }
  )
), v = /* @__PURE__ */ e(
  r(
    "arcgis-feature-table",
    {
      onArcgisCellClick: "arcgisCellClick",
      onArcgisCellKeydown: "arcgisCellKeydown",
      onArcgisCellPointerout: "arcgisCellPointerout",
      onArcgisCellPointerover: "arcgisCellPointerover",
      onArcgisColumnReorder: "arcgisColumnReorder",
      onArcgisPropertyChange: "arcgisPropertyChange",
      onArcgisReady: "arcgisReady",
      onArcgisSelectionChange: "arcgisSelectionChange"
    }
  )
), E = /* @__PURE__ */ e(
  r(
    "arcgis-feature-templates-legacy",
    {
      onArcgisPropertyChange: "arcgisPropertyChange",
      onArcgisReady: "arcgisReady",
      onArcgisSelect: "arcgisSelect"
    }
  )
), U = /* @__PURE__ */ e(
  r(
    "arcgis-features",
    {
      onArcgisClose: "arcgisClose",
      onArcgisPropertyChange: "arcgisPropertyChange",
      onArcgisReady: "arcgisReady",
      onArcgisTriggerAction: "arcgisTriggerAction"
    }
  )
), F = /* @__PURE__ */ e(
  r(
    "arcgis-floor-filter",
    {
      onArcgisReady: "arcgisReady"
    }
  )
), f = /* @__PURE__ */ e(
  r(
    "arcgis-fullscreen",
    {
      onArcgisPropertyChange: "arcgisPropertyChange",
      onArcgisReady: "arcgisReady"
    }
  )
), M = /* @__PURE__ */ e(
  r(
    "arcgis-grid-controls",
    {
      onArcgisPropertyChange: "arcgisPropertyChange",
      onArcgisReady: "arcgisReady"
    }
  )
), I = /* @__PURE__ */ e(
  r(
    "arcgis-histogram-range-slider",
    {
      onArcgisPropertyChange: "arcgisPropertyChange",
      onArcgisReady: "arcgisReady"
    }
  )
), B = /* @__PURE__ */ e(
  r(
    "arcgis-home",
    {
      onArcgisGo: "arcgisGo",
      onArcgisPropertyChange: "arcgisPropertyChange",
      onArcgisReady: "arcgisReady"
    }
  )
), G = /* @__PURE__ */ e(
  r(
    "arcgis-layer-list",
    {
      onArcgisClose: "arcgisClose",
      onArcgisPropertyChange: "arcgisPropertyChange",
      onArcgisReady: "arcgisReady",
      onArcgisTriggerAction: "arcgisTriggerAction"
    }
  )
), N = /* @__PURE__ */ e(
  r(
    "arcgis-legend",
    {
      onArcgisPropertyChange: "arcgisPropertyChange",
      onArcgisReady: "arcgisReady"
    }
  )
), H = /* @__PURE__ */ e(
  r(
    "arcgis-line-of-sight",
    {
      onArcgisPropertyChange: "arcgisPropertyChange",
      onArcgisReady: "arcgisReady"
    }
  )
), K = /* @__PURE__ */ e(
  r(
    "arcgis-link-chart",
    {
      onArcgisLoadError: "arcgisLoadError",
      onArcgisViewAnalysisViewCreate: "arcgisViewAnalysisViewCreate",
      onArcgisViewAnalysisViewCreateError: "arcgisViewAnalysisViewCreateError",
      onArcgisViewAnalysisViewDestroy: "arcgisViewAnalysisViewDestroy",
      onArcgisViewChange: "arcgisViewChange",
      onArcgisViewClick: "arcgisViewClick",
      onArcgisViewDoubleClick: "arcgisViewDoubleClick",
      onArcgisViewDoubleTapDrag: "arcgisViewDoubleTapDrag",
      onArcgisViewDrag: "arcgisViewDrag",
      onArcgisViewHold: "arcgisViewHold",
      onArcgisViewImmediateClick: "arcgisViewImmediateClick",
      onArcgisViewImmediateDoubleClick: "arcgisViewImmediateDoubleClick",
      onArcgisViewKeyDown: "arcgisViewKeyDown",
      onArcgisViewKeyUp: "arcgisViewKeyUp",
      onArcgisViewLayerviewCreate: "arcgisViewLayerviewCreate",
      onArcgisViewLayerviewCreateError: "arcgisViewLayerviewCreateError",
      onArcgisViewLayerviewDestroy: "arcgisViewLayerviewDestroy",
      onArcgisViewMouseWheel: "arcgisViewMouseWheel",
      onArcgisViewPointerDown: "arcgisViewPointerDown",
      onArcgisViewPointerEnter: "arcgisViewPointerEnter",
      onArcgisViewPointerLeave: "arcgisViewPointerLeave",
      onArcgisViewPointerMove: "arcgisViewPointerMove",
      onArcgisViewPointerUp: "arcgisViewPointerUp",
      onArcgisViewReadyChange: "arcgisViewReadyChange",
      onArcgisViewReadyError: "arcgisViewReadyError",
      onArcgisViewVerticalTwoFingerDrag: "arcgisViewVerticalTwoFingerDrag"
    }
  )
), W = /* @__PURE__ */ e(
  r(
    "arcgis-link-chart-layout-switcher",
    {
      onArcgisPropertyChange: "arcgisPropertyChange",
      onArcgisReady: "arcgisReady",
      onArcgisSwitchLayout: "arcgisSwitchLayout"
    }
  )
), z = /* @__PURE__ */ e(
  r(
    "arcgis-link-chart-nonspatial-visibility-toggle",
    {
      onArcgisPropertyChange: "arcgisPropertyChange",
      onArcgisReady: "arcgisReady"
    }
  )
), x = /* @__PURE__ */ e(
  r(
    "arcgis-locate",
    {
      onArcgisError: "arcgisError",
      onArcgisPropertyChange: "arcgisPropertyChange",
      onArcgisReady: "arcgisReady",
      onArcgisSuccess: "arcgisSuccess"
    }
  )
), O = /* @__PURE__ */ e(
  r(
    "arcgis-map",
    {
      onArcgisLoadError: "arcgisLoadError",
      onArcgisViewAnalysisViewCreate: "arcgisViewAnalysisViewCreate",
      onArcgisViewAnalysisViewCreateError: "arcgisViewAnalysisViewCreateError",
      onArcgisViewAnalysisViewDestroy: "arcgisViewAnalysisViewDestroy",
      onArcgisViewChange: "arcgisViewChange",
      onArcgisViewClick: "arcgisViewClick",
      onArcgisViewDoubleClick: "arcgisViewDoubleClick",
      onArcgisViewDoubleTapDrag: "arcgisViewDoubleTapDrag",
      onArcgisViewDrag: "arcgisViewDrag",
      onArcgisViewHold: "arcgisViewHold",
      onArcgisViewImmediateClick: "arcgisViewImmediateClick",
      onArcgisViewImmediateDoubleClick: "arcgisViewImmediateDoubleClick",
      onArcgisViewKeyDown: "arcgisViewKeyDown",
      onArcgisViewKeyUp: "arcgisViewKeyUp",
      onArcgisViewLayerviewCreate: "arcgisViewLayerviewCreate",
      onArcgisViewLayerviewCreateError: "arcgisViewLayerviewCreateError",
      onArcgisViewLayerviewDestroy: "arcgisViewLayerviewDestroy",
      onArcgisViewMouseWheel: "arcgisViewMouseWheel",
      onArcgisViewPointerDown: "arcgisViewPointerDown",
      onArcgisViewPointerEnter: "arcgisViewPointerEnter",
      onArcgisViewPointerLeave: "arcgisViewPointerLeave",
      onArcgisViewPointerMove: "arcgisViewPointerMove",
      onArcgisViewPointerUp: "arcgisViewPointerUp",
      onArcgisViewReadyChange: "arcgisViewReadyChange",
      onArcgisViewReadyError: "arcgisViewReadyError",
      onArcgisViewVerticalTwoFingerDrag: "arcgisViewVerticalTwoFingerDrag"
    }
  )
), Z = /* @__PURE__ */ e(
  r(
    "arcgis-measurement",
    {
      onArcgisPropertyChange: "arcgisPropertyChange",
      onArcgisReady: "arcgisReady"
    }
  )
), j = /* @__PURE__ */ e(
  r(
    "arcgis-navigation-toggle",
    {
      onArcgisPropertyChange: "arcgisPropertyChange",
      onArcgisReady: "arcgisReady"
    }
  )
), q = /* @__PURE__ */ e(
  r(
    "arcgis-oriented-imagery-viewer",
    {
      onArcgisPropertyChange: "arcgisPropertyChange",
      onArcgisReady: "arcgisReady"
    }
  )
), J = /* @__PURE__ */ e(
  r(
    "arcgis-paste",
    {
      onArcgisApplyPasteCommand: "arcgisApplyPasteCommand",
      onArcgisApplyPasteError: "arcgisApplyPasteError",
      onArcgisBeforeClipboardMove: "arcgisBeforeClipboardMove",
      onArcgisClipboardMoveStateChanged: "arcgisClipboardMoveStateChanged",
      onArcgisClose: "arcgisClose",
      onArcgisPropertyChange: "arcgisPropertyChange",
      onArcgisReady: "arcgisReady"
    }
  )
), Q = /* @__PURE__ */ e(
  r(
    "arcgis-clipboard",
    {
      onArcgisReady: "arcgisReady"
    }
  )
), X = /* @__PURE__ */ e(
  r(
    "arcgis-feature-image",
    {}
  )
), Y = /* @__PURE__ */ e(
  r(
    "arcgis-feature-title",
    {}
  )
), _ = /* @__PURE__ */ e(
  r(
    "arcgis-multi-feature-form",
    {
      onArcgisPropertyChange: "arcgisPropertyChange",
      onArcgisReady: "arcgisReady",
      onArcgisSubmit: "arcgisSubmit",
      onArcgisValueChange: "arcgisValueChange"
    }
  )
), $ = /* @__PURE__ */ e(
  r(
    "arcgis-popup",
    {
      onArcgisClose: "arcgisClose",
      onArcgisPropertyChange: "arcgisPropertyChange",
      onArcgisReady: "arcgisReady",
      onArcgisTriggerAction: "arcgisTriggerAction"
    }
  )
), rr = /* @__PURE__ */ e(
  r(
    "arcgis-presentation",
    {
      onArcgisReady: "arcgisReady"
    }
  )
), er = /* @__PURE__ */ e(
  r(
    "arcgis-print",
    {
      onArcgisBrowseTemplatesChange: "arcgisBrowseTemplatesChange",
      onArcgisComplete: "arcgisComplete",
      onArcgisPropertyChange: "arcgisPropertyChange",
      onArcgisReady: "arcgisReady",
      onArcgisSaveAs: "arcgisSaveAs",
      onArcgisSubmit: "arcgisSubmit"
    }
  )
), ir = /* @__PURE__ */ e(
  r(
    "arcgis-reverse-geocode-list",
    {
      onArcgisReverseGeocodeResultsChange: "arcgisReverseGeocodeResultsChange"
    }
  )
), ar = /* @__PURE__ */ e(
  r(
    "arcgis-scale-bar",
    {
      onArcgisReady: "arcgisReady"
    }
  )
), gr = /* @__PURE__ */ e(
  r(
    "arcgis-scale-range-slider",
    {
      onArcgisPropertyChange: "arcgisPropertyChange",
      onArcgisReady: "arcgisReady"
    }
  )
), sr = /* @__PURE__ */ e(
  r(
    "arcgis-scene",
    {
      onArcgisLoadError: "arcgisLoadError",
      onArcgisViewAnalysisViewCreate: "arcgisViewAnalysisViewCreate",
      onArcgisViewAnalysisViewCreateError: "arcgisViewAnalysisViewCreateError",
      onArcgisViewAnalysisViewDestroy: "arcgisViewAnalysisViewDestroy",
      onArcgisViewChange: "arcgisViewChange",
      onArcgisViewClick: "arcgisViewClick",
      onArcgisViewDoubleClick: "arcgisViewDoubleClick",
      onArcgisViewDoubleTapDrag: "arcgisViewDoubleTapDrag",
      onArcgisViewDrag: "arcgisViewDrag",
      onArcgisViewHold: "arcgisViewHold",
      onArcgisViewImmediateClick: "arcgisViewImmediateClick",
      onArcgisViewImmediateDoubleClick: "arcgisViewImmediateDoubleClick",
      onArcgisViewKeyDown: "arcgisViewKeyDown",
      onArcgisViewKeyUp: "arcgisViewKeyUp",
      onArcgisViewLayerviewCreate: "arcgisViewLayerviewCreate",
      onArcgisViewLayerviewCreateError: "arcgisViewLayerviewCreateError",
      onArcgisViewLayerviewDestroy: "arcgisViewLayerviewDestroy",
      onArcgisViewMouseWheel: "arcgisViewMouseWheel",
      onArcgisViewPointerDown: "arcgisViewPointerDown",
      onArcgisViewPointerEnter: "arcgisViewPointerEnter",
      onArcgisViewPointerLeave: "arcgisViewPointerLeave",
      onArcgisViewPointerMove: "arcgisViewPointerMove",
      onArcgisViewPointerUp: "arcgisViewPointerUp",
      onArcgisViewReadyChange: "arcgisViewReadyChange",
      onArcgisViewReadyError: "arcgisViewReadyError",
      onArcgisViewVerticalTwoFingerDrag: "arcgisViewVerticalTwoFingerDrag"
    }
  )
), cr = /* @__PURE__ */ e(
  r(
    "arcgis-search",
    {
      onArcgisPropertyChange: "arcgisPropertyChange",
      onArcgisReady: "arcgisReady",
      onArcgisSearchClear: "arcgisSearchClear",
      onArcgisSearchComplete: "arcgisSearchComplete",
      onArcgisSearchStart: "arcgisSearchStart",
      onArcgisSelectResult: "arcgisSelectResult",
      onArcgisSuggestComplete: "arcgisSuggestComplete",
      onArcgisSuggestStart: "arcgisSuggestStart"
    }
  )
), or = /* @__PURE__ */ e(
  r(
    "arcgis-search-result-renderer",
    {
      onArcgisReady: "arcgisReady"
    }
  )
), nr = /* @__PURE__ */ e(
  r(
    "arcgis-shadow-cast",
    {
      onArcgisPropertyChange: "arcgisPropertyChange",
      onArcgisReady: "arcgisReady"
    }
  )
), tr = /* @__PURE__ */ e(
  r(
    "arcgis-sketch",
    {
      onArcgisCreate: "arcgisCreate",
      onArcgisDelete: "arcgisDelete",
      onArcgisPropertyChange: "arcgisPropertyChange",
      onArcgisReady: "arcgisReady",
      onArcgisRedo: "arcgisRedo",
      onArcgisUndo: "arcgisUndo",
      onArcgisUpdate: "arcgisUpdate"
    }
  )
), Ar = /* @__PURE__ */ e(
  r(
    "arcgis-slice",
    {
      onArcgisPropertyChange: "arcgisPropertyChange",
      onArcgisReady: "arcgisReady"
    }
  )
), yr = /* @__PURE__ */ e(
  r(
    "arcgis-slider-binary-color-size-legacy",
    {
      onArcgisPropertyChange: "arcgisPropertyChange",
      onArcgisReady: "arcgisReady",
      onArcgisThumbChange: "arcgisThumbChange",
      onArcgisThumbDrag: "arcgisThumbDrag"
    }
  )
), lr = /* @__PURE__ */ e(
  r(
    "arcgis-slider-classed-color-legacy",
    {
      onArcgisPropertyChange: "arcgisPropertyChange",
      onArcgisReady: "arcgisReady",
      onArcgisThumbChange: "arcgisThumbChange",
      onArcgisThumbDrag: "arcgisThumbDrag"
    }
  )
), dr = /* @__PURE__ */ e(
  r(
    "arcgis-slider-classed-size-legacy",
    {
      onArcgisPropertyChange: "arcgisPropertyChange",
      onArcgisReady: "arcgisReady",
      onArcgisThumbChange: "arcgisThumbChange",
      onArcgisThumbDrag: "arcgisThumbDrag"
    }
  )
), Cr = /* @__PURE__ */ e(
  r(
    "arcgis-slider-color-legacy",
    {
      onArcgisPropertyChange: "arcgisPropertyChange",
      onArcgisReady: "arcgisReady",
      onArcgisThumbChange: "arcgisThumbChange",
      onArcgisThumbDrag: "arcgisThumbDrag"
    }
  )
), hr = /* @__PURE__ */ e(
  r(
    "arcgis-slider-color-size-legacy",
    {
      onArcgisPropertyChange: "arcgisPropertyChange",
      onArcgisReady: "arcgisReady",
      onArcgisThumbChange: "arcgisThumbChange",
      onArcgisThumbDrag: "arcgisThumbDrag"
    }
  )
), pr = /* @__PURE__ */ e(
  r(
    "arcgis-slider-heatmap-legacy",
    {
      onArcgisPropertyChange: "arcgisPropertyChange",
      onArcgisReady: "arcgisReady",
      onArcgisThumbChange: "arcgisThumbChange",
      onArcgisThumbDrag: "arcgisThumbDrag"
    }
  )
), wr = /* @__PURE__ */ e(
  r(
    "arcgis-slider-opacity-legacy",
    {
      onArcgisPropertyChange: "arcgisPropertyChange",
      onArcgisReady: "arcgisReady",
      onArcgisThumbChange: "arcgisThumbChange",
      onArcgisThumbDrag: "arcgisThumbDrag"
    }
  )
), Rr = /* @__PURE__ */ e(
  r(
    "arcgis-slider-size-legacy",
    {
      onArcgisPropertyChange: "arcgisPropertyChange",
      onArcgisReady: "arcgisReady",
      onArcgisThumbChange: "arcgisThumbChange",
      onArcgisThumbDrag: "arcgisThumbDrag"
    }
  )
), Vr = /* @__PURE__ */ e(
  r(
    "arcgis-snapping-controls",
    {
      onArcgisPropertyChange: "arcgisPropertyChange",
      onArcgisReady: "arcgisReady"
    }
  )
), Pr = /* @__PURE__ */ e(
  r(
    "arcgis-color-select",
    {
      onArcgisColorChanged: "arcgisColorChanged"
    }
  )
), ur = /* @__PURE__ */ e(
  r(
    "arcgis-time-of-day-slider",
    {
      onArcgisTimeChange: "arcgisTimeChange"
    }
  )
), mr = /* @__PURE__ */ e(
  r(
    "arcgis-webgl-error",
    {}
  )
), Tr = /* @__PURE__ */ e(
  r(
    "arcgis-swipe",
    {
      onArcgisPropertyChange: "arcgisPropertyChange",
      onArcgisReady: "arcgisReady",
      onArcgisSwipeChange: "arcgisSwipeChange",
      onArcgisSwipeInput: "arcgisSwipeInput"
    }
  )
), Sr = /* @__PURE__ */ e(
  r(
    "arcgis-table-list",
    {
      onArcgisClose: "arcgisClose",
      onArcgisPropertyChange: "arcgisPropertyChange",
      onArcgisReady: "arcgisReady",
      onArcgisTriggerAction: "arcgisTriggerAction"
    }
  )
), Dr = /* @__PURE__ */ e(
  r(
    "arcgis-time-slider",
    {
      onArcgisPropertyChange: "arcgisPropertyChange",
      onArcgisReady: "arcgisReady",
      onArcgisTriggerAction: "arcgisTriggerAction"
    }
  )
), br = /* @__PURE__ */ e(
  r(
    "arcgis-time-zone-label",
    {
      onArcgisPropertyChange: "arcgisPropertyChange",
      onArcgisReady: "arcgisReady"
    }
  )
), Lr = /* @__PURE__ */ e(
  r(
    "arcgis-track",
    {
      onArcgisComplete: "arcgisComplete",
      onArcgisError: "arcgisError",
      onArcgisPropertyChange: "arcgisPropertyChange",
      onArcgisReady: "arcgisReady"
    }
  )
), kr = /* @__PURE__ */ e(
  r(
    "arcgis-utility-network-associations",
    {
      onArcgisReady: "arcgisReady"
    }
  )
), vr = /* @__PURE__ */ e(
  r(
    "arcgis-utility-network-trace",
    {
      onArcgisAddFlag: "arcgisAddFlag",
      onArcgisAddFlagComplete: "arcgisAddFlagComplete",
      onArcgisAddFlagError: "arcgisAddFlagError",
      onArcgisAddResultArea: "arcgisAddResultArea",
      onArcgisPropertyChange: "arcgisPropertyChange",
      onArcgisReady: "arcgisReady",
      onArcgisRemoveResultArea: "arcgisRemoveResultArea"
    }
  )
), Er = /* @__PURE__ */ e(
  r(
    "arcgis-utility-network-trace-analysis",
    {
      onArcgisHideGraphicsChange: "arcgisHideGraphicsChange",
      onArcgisReady: "arcgisReady",
      onArcgisTraceLocationSelectEnd: "arcgisTraceLocationSelectEnd",
      onArcgisTraceLocationSelectStart: "arcgisTraceLocationSelectStart",
      onArcgisTraceResultClear: "arcgisTraceResultClear",
      onArcgisTraceResultFeaturesReady: "arcgisTraceResultFeaturesReady"
    }
  )
), Ur = /* @__PURE__ */ e(
  r(
    "arcgis-utility-network-aggregated-geometry",
    {
      onArcgisHideGraphicsChange: "arcgisHideGraphicsChange"
    }
  )
), Fr = /* @__PURE__ */ e(
  r(
    "arcgis-utility-network-subnetwork-info",
    {}
  )
), fr = /* @__PURE__ */ e(
  r(
    "arcgis-utility-network-trace-inputs",
    {
      onArcgisPropertyChange: "arcgisPropertyChange",
      onArcgisSelectedNamedTraceConfigurationChange: "arcgisSelectedNamedTraceConfigurationChange",
      onArcgisTraceInputsChange: "arcgisTraceInputsChange",
      onArcgisTraceLocationSelectEnd: "arcgisTraceLocationSelectEnd",
      onArcgisTraceLocationSelectStart: "arcgisTraceLocationSelectStart",
      onArcgisUtilityNetworkChange: "arcgisUtilityNetworkChange"
    }
  )
), Mr = /* @__PURE__ */ e(
  r(
    "arcgis-utility-network-trace-location",
    {
      onArcgisTraceLocationChange: "arcgisTraceLocationChange",
      onArcgisTraceLocationSelectEnd: "arcgisTraceLocationSelectEnd",
      onArcgisTraceLocationSelectStart: "arcgisTraceLocationSelectStart"
    }
  )
), Ir = /* @__PURE__ */ e(
  r(
    "arcgis-utility-network-trace-location-list",
    {}
  )
), Br = /* @__PURE__ */ e(
  r(
    "arcgis-utility-network-trace-results",
    {
      onArcgisAggregatedGeometryGraphicsColorChange: "arcgisAggregatedGeometryGraphicsColorChange",
      onArcgisHideGraphicsChange: "arcgisHideGraphicsChange",
      onArcgisSelectedLayerChange: "arcgisSelectedLayerChange",
      onArcgisSelectedResultTypeChange: "arcgisSelectedResultTypeChange",
      onArcgisTraceResultFeaturesReady: "arcgisTraceResultFeaturesReady"
    }
  )
), Gr = /* @__PURE__ */ e(
  r(
    "arcgis-utility-network-validate-topology",
    {
      onArcgisPropertyChange: "arcgisPropertyChange",
      onArcgisReady: "arcgisReady"
    }
  )
), Nr = /* @__PURE__ */ e(
  r(
    "arcgis-value-picker-legacy",
    {
      onArcgisAnimate: "arcgisAnimate",
      onArcgisNext: "arcgisNext",
      onArcgisPause: "arcgisPause",
      onArcgisPlay: "arcgisPlay",
      onArcgisPrevious: "arcgisPrevious",
      onArcgisPropertyChange: "arcgisPropertyChange",
      onArcgisReady: "arcgisReady"
    }
  )
), Hr = /* @__PURE__ */ e(
  r(
    "arcgis-version-management",
    {
      onArcgisReady: "arcgisReady",
      onArcgisVersioningStateChanged: "arcgisVersioningStateChanged"
    }
  )
), Kr = /* @__PURE__ */ e(
  r(
    "arcgis-video",
    {
      onArcgisViewReadyChange: "arcgisViewReadyChange"
    }
  )
), Wr = /* @__PURE__ */ e(
  r(
    "arcgis-video-player",
    {
      onArcgisPropertyChange: "arcgisPropertyChange",
      onArcgisReady: "arcgisReady",
      onArcgisVideoReady: "arcgisVideoReady"
    }
  )
), zr = /* @__PURE__ */ e(
  r(
    "arcgis-volume-measurement",
    {
      onArcgisPropertyChange: "arcgisPropertyChange",
      onArcgisReady: "arcgisReady"
    }
  )
), xr = /* @__PURE__ */ e(
  r(
    "arcgis-weather",
    {
      onArcgisPropertyChange: "arcgisPropertyChange",
      onArcgisReady: "arcgisReady"
    }
  )
), Or = /* @__PURE__ */ e(
  r(
    "arcgis-zoom",
    {
      onArcgisActionZoom: "arcgisActionZoom",
      onArcgisPropertyChange: "arcgisPropertyChange",
      onArcgisReady: "arcgisReady"
    }
  )
);
export {
  n as ArcgisAreaMeasurement2d,
  t as ArcgisAreaMeasurement3d,
  A as ArcgisAttachments,
  y as ArcgisBasemapGallery,
  l as ArcgisBasemapLayerList,
  d as ArcgisBasemapToggle,
  C as ArcgisBookmarks,
  h as ArcgisBuildingExplorer,
  p as ArcgisCatalogLayerList,
  Q as ArcgisClipboard,
  Pr as ArcgisColorSelect,
  w as ArcgisCompass,
  R as ArcgisCoordinateConversion,
  V as ArcgisDaylight,
  m as ArcgisDirectLineMeasurement3d,
  P as ArcgisDirectionalPad,
  u as ArcgisDirections,
  T as ArcgisDistanceMeasurement2d,
  S as ArcgisEditor,
  D as ArcgisElevationProfile,
  b as ArcgisExpand,
  L as ArcgisFeature,
  k as ArcgisFeatureForm,
  X as ArcgisFeatureImage,
  v as ArcgisFeatureTable,
  E as ArcgisFeatureTemplatesLegacy,
  Y as ArcgisFeatureTitle,
  U as ArcgisFeatures,
  F as ArcgisFloorFilter,
  f as ArcgisFullscreen,
  M as ArcgisGridControls,
  I as ArcgisHistogramRangeSlider,
  B as ArcgisHome,
  G as ArcgisLayerList,
  N as ArcgisLegend,
  H as ArcgisLineOfSight,
  K as ArcgisLinkChart,
  W as ArcgisLinkChartLayoutSwitcher,
  z as ArcgisLinkChartNonspatialVisibilityToggle,
  x as ArcgisLocate,
  O as ArcgisMap,
  Z as ArcgisMeasurement,
  _ as ArcgisMultiFeatureForm,
  j as ArcgisNavigationToggle,
  q as ArcgisOrientedImageryViewer,
  J as ArcgisPaste,
  $ as ArcgisPopup,
  rr as ArcgisPresentation,
  er as ArcgisPrint,
  ir as ArcgisReverseGeocodeList,
  ar as ArcgisScaleBar,
  gr as ArcgisScaleRangeSlider,
  sr as ArcgisScene,
  cr as ArcgisSearch,
  or as ArcgisSearchResultRenderer,
  nr as ArcgisShadowCast,
  tr as ArcgisSketch,
  Ar as ArcgisSlice,
  yr as ArcgisSliderBinaryColorSizeLegacy,
  lr as ArcgisSliderClassedColorLegacy,
  dr as ArcgisSliderClassedSizeLegacy,
  Cr as ArcgisSliderColorLegacy,
  hr as ArcgisSliderColorSizeLegacy,
  pr as ArcgisSliderHeatmapLegacy,
  wr as ArcgisSliderOpacityLegacy,
  Rr as ArcgisSliderSizeLegacy,
  Vr as ArcgisSnappingControls,
  Tr as ArcgisSwipe,
  Sr as ArcgisTableList,
  ur as ArcgisTimeOfDaySlider,
  Dr as ArcgisTimeSlider,
  br as ArcgisTimeZoneLabel,
  Lr as ArcgisTrack,
  Ur as ArcgisUtilityNetworkAggregatedGeometry,
  kr as ArcgisUtilityNetworkAssociations,
  Fr as ArcgisUtilityNetworkSubnetworkInfo,
  vr as ArcgisUtilityNetworkTrace,
  Er as ArcgisUtilityNetworkTraceAnalysis,
  fr as ArcgisUtilityNetworkTraceInputs,
  Mr as ArcgisUtilityNetworkTraceLocation,
  Ir as ArcgisUtilityNetworkTraceLocationList,
  Br as ArcgisUtilityNetworkTraceResults,
  Gr as ArcgisUtilityNetworkValidateTopology,
  Nr as ArcgisValuePickerLegacy,
  Hr as ArcgisVersionManagement,
  Kr as ArcgisVideo,
  Wr as ArcgisVideoPlayer,
  zr as ArcgisVolumeMeasurement,
  xr as ArcgisWeather,
  mr as ArcgisWebglError,
  Or as ArcgisZoom
};
