// @arcgis/map-components/dist/components/arcgis-print-exports-panel/customElement.js
/* COPYRIGHT Esri - https://js.arcgis.com/5.0/LICENSE.txt */
import { c as p } from "../../chunks/runtime.js";
import { css as f, html as s } from "lit";
import { whenOnce as h } from "@arcgis/core/core/reactiveUtils.js";
import { addProxy as g } from "@arcgis/core/core/urlUtils.js";
import { LitElement as v, safeClassMap as u, nothing as x } from "@arcgis/lumina";
import { classes as $ } from "@arcgis/toolkit/dom";
import { hasSameOrigin as w } from "@arcgis/toolkit/url";
const b = f`.export-section{padding:var(--calcite-spacing-xxs)}.export-section-centered{text-align:center}.exported-files-empty{display:flex;flex-direction:column;align-items:center;padding:var(--calcite-spacing-lg) var(--calcite-spacing-sm);font-size:small}.exported-files-title{font-weight:700;font-size:var(--calcite-font-size)}.exported-file{display:flex;align-items:center;justify-content:space-between}.exported-file-link-title,.exported-file-link-description{word-wrap:break-word;word-break:break-all;overflow:hidden;text-overflow:ellipsis;display:-webkit-box;-webkit-line-clamp:2;line-clamp:2;-webkit-box-orient:vertical}`;
function k(c) {
  const { state: e, extension: t } = c;
  switch (e) {
    case "pending":
      return "spinner";
    case "error":
      return "exclamation-mark-circle";
    default:
      return t?.toLowerCase()?.includes("pdf") ? "file-pdf" : "file";
  }
}
function d(c, e) {
  return !c || c === "all" || !!(e && c.includes(e));
}
class E extends v {
  constructor() {
    super(...arguments), this._handleLinkClick = (e) => {
      if (e?.state !== "ready" || !e.url)
        return;
      const t = e.url, i = document.createElement("a");
      i.target = "_blank", i.href = t, i.rel = "noreferrer", i.download = e.formattedName ?? "";
      {
        i.click();
        return;
      }
    };
  }
  static {
    this.properties = { showSaveAs: 5, viewModel: 0, messages: 0, allowedFormatsForSaving: 1, arcgisSaveAs: 0 };
  }
  static {
    this.styles = b;
  }
  loaded() {
    this.manager.onLifecycle(() => [this.viewModel.exportedLinks.on("change", () => this.requestUpdate())]);
  }
  render() {
    const e = this.viewModel.exportedLinks, t = {
      "export-section-centered": !e.length
    };
    return s`<section class=${u($("export-section", t))}>${e.length === 0 ? s`<div class="exported-files-empty"><calcite-icon icon=file scale=l></calcite-icon><div><p class="exported-files-title">${this.messages.noExportedFiles}</p><div>${this.messages.exportHint}</div></div></div>` : s`<calcite-list .filterEnabled=${e.length > 10} .label=${this.messages.exportedLinksList} selection-mode=none>${e.map((i) => s`<calcite-list-item .label=${i.formattedName ?? ""}>${this._renderExportedLinkItemContent(i)}${this._renderExportedLinkItemIcon(i)}${this._renderExportedLinkItemAction(i)}</calcite-list-item>${this._renderExportedLinkItemPopover(i)}`)}</calcite-list>`}</section>`;
  }
  _renderExportedLinkItemPopover(e) {
    const t = this.messages, i = e.formattedName ?? "", { state: l } = e, a = this.showSaveAs && l !== "error" && d(this.allowedFormatsForSaving, e.extension);
    return l !== "pending" && a ? s`<calcite-popover auto-close label=Popover offset-distance=1 overlay-positioning=fixed pointer-disabled .referenceElement=${`${i}-popover-button`} scale=s><calcite-list label>${e.portalItem ? s`<calcite-list-item .label=${t.openItem} @calciteListItemSelect=${() => {
      const o = e.portalItem;
      if (!o?.itemPageUrl)
        return;
      const n = document.createElement("a");
      n.target = "_blank", n.href = o.itemPageUrl, n.rel = "noreferrer", n.click();
    }}><calcite-icon icon=launch2 scale=s slot=content-start></calcite-icon></calcite-list-item>` : s`<calcite-list-item .label=${t.saveAs} @calciteListItemSelect=${async () => {
      h(() => e.portalItem).then((o) => {
        o.itemPageUrl && this.requestUpdate();
      }), this.arcgisSaveAs.emit(e);
    }}><calcite-icon icon=save scale=s slot=content-start></calcite-icon></calcite-list-item>`}<calcite-list-item label=Delete @calciteListItemSelect=${() => this.viewModel.exportedLinks.remove(e)}><calcite-icon icon=trash scale=s slot=content-start></calcite-icon></calcite-list-item></calcite-list></calcite-popover>` : null;
  }
  _renderExportedLinkItemContent(e) {
    const t = this.messages, { url: i, state: l, error: a } = e, o = e.formattedName ?? "";
    let n = i || null;
    n && (n = g(n));
    const r = w(n, document.location.href), m = l === "error" ? a?.name === "print-task:cim-symbol-unsupported" ? t.exportWebMapCIMError : t.exportWebMapError : null;
    return s`<div class="exported-file" slot=content @click=${() => this._handleLinkClick(e)}><div><div class="exported-file-link-title">${o}</div><div class="exported-file-link-description">${l === "pending" ? t.generatingExport : m || (r ? t.ready : t.linkReady)}</div></div>${l === "ready" ? s`<calcite-icon .ariaLabel=${`${o}. ${t.linkReady}`} .icon=${r ? "download-to" : "launch"} scale=s slot=content-end></calcite-icon>` : null}</div>`;
  }
  _renderExportedLinkItemIcon(e) {
    return e.state === "pending" ? s`<calcite-loader inline .label=${this.messages.generatingExport} scale=m slot=content-start></calcite-loader>` : s`<calcite-icon .icon=${k(e)} scale=s slot=content-start></calcite-icon>`;
  }
  _renderExportedLinkItemAction(e) {
    const { state: t } = e, i = e.formattedName ?? "", l = this.showSaveAs && t !== "error" && d(this.allowedFormatsForSaving, e.extension);
    return t !== "pending" ? s`<calcite-action .icon=${l ? "ellipsis" : "x"} id=${(l ? `${i}-popover-button` : void 0) ?? x} @click=${l ? void 0 : () => {
      this.viewModel.exportedLinks.remove(e);
    }} slot=actions-end .text=${l ? "options" : "remove"}></calcite-action>` : null;
  }
}
p("arcgis-print-exports-panel", E);
export {
  E as PrintExportsPanel
};
