// @arcgis/map-components/dist/components/arcgis-version-management-service-item/customElement.js
/* COPYRIGHT Esri - https://js.arcgis.com/5.0/LICENSE.txt */
import { c as M } from "../../chunks/runtime.js";
import { LitElement as A, createEvent as o, safeClassMap as v } from "@arcgis/lumina";
import { html as t, render as U } from "lit";
class F extends A {
  constructor() {
    super(...arguments), this._showInvalidVersionNotice = !0, this.arcgisCreateVersionList = o(), this.arcgisFlowItemBack = o(), this.arcgisFlowItemClose = o(), this.arcgisGetVersions = o(), this.arcgisManageVersion = o(), this.arcgisNewVersion = o();
  }
  static {
    this.properties = { _showInvalidVersionNotice: 16, serviceItemElementProps: 0 };
  }
  disconnectedCallback() {
    super.disconnectedCallback(), this.versionListContainer?.remove();
  }
  render() {
    const { serviceItemElementProps: { allowEditing: h, closable: $, currentUser: p, executionError: V, currentVersionIdentifier: r, flowElement: b, heading: I, isInitialVersionInvalid: f, isVersionAdministrator: u, isVersioningApiAvailable: w, mode: N, pageSize: x, serviceName: i, serviceUrl: c, state: E, strings: n, versionInfos: a } } = this;
    let { _versionList: l } = this;
    const C = f.get(c) && this._showInvalidVersionNotice ? t`<calcite-notice class=${v(this._showInvalidVersionNotice ? "invalid-initial-version-notice" : "")} closable kind=warning open scale=s width=full icon @calciteNoticeClose=${() => this._showInvalidVersionNotice = !1}><div slot=title>${n.headers.invalidInitialVersion}</div><div slot=message>${n.loadErrors.invalidInitialVersion}</div></calcite-notice>` : void 0, L = t`<calcite-combobox clear-disabled .ariaLabel=${`combobox-${i}`} .label=${`combobox-${i}`} max-items=5 overlay-positioning=fixed selection-mode=single-persist @calciteComboboxChange=${(s) => {
      const m = s.target, e = a.find((k) => k.versionIdentifier.guid === m.value);
      this.arcgisManageVersion.emit({ actionType: "changeVersion", serviceUrl: c, versionInfo: e });
    }}>${a ? a.map((s) => t`<calcite-combobox-item class=${v(N === "dialog" ? "" : "calcite-combobox-item")} .heading=${s.versionIdentifier.name} .value=${s.versionIdentifier.guid} .selected=${r.guid === s.versionIdentifier.guid && r.name === s.versionIdentifier.name}></calcite-combobox-item>`) : void 0}</calcite-combobox>`;
    let d, g;
    return h && (d = t`<calcite-action icon=list .id=${`action-manage-versions-${i}`} .label=${`action-manage-versions-${i}`} scale=s slot=actions-end .text=${`action-manage-versions-${i}`} @click=${async () => {
      const s = {
        closable: $,
        currentUser: p,
        currentVersionIdentifier: r,
        executionError: V,
        heading: I,
        isVersionAdministrator: u,
        isVersioningApiAvailable: w,
        pageSize: x,
        serviceName: i,
        serviceUrl: c,
        state: E,
        strings: n,
        versionInfos: a
      };
      this.versionListContainer = document.createElement("div");
      const m = U(t`<arcgis-version-management-version-list .versionListElementProps=${s} selected show-back-button @arcgisFlowItemBack=${(e) => {
        e.preventDefault(), this.arcgisFlowItemBack.emit();
      }} @arcgisFlowItemClose=${() => this.arcgisFlowItemClose.emit()} @arcgisGetVersions=${(e) => this.arcgisGetVersions.emit({ serviceUrl: e.detail.serviceUrl })} @arcgisManageVersion=${(e) => this.arcgisManageVersion.emit(e.detail)} @arcgisNewVersion=${(e) => this.arcgisNewVersion.emit({ serviceUrl: e.detail.serviceUrl })}></arcgis-version-management-version-list>`, this.versionListContainer);
      l = Array.from(m.parentNode.childNodes).find((e) => e.nodeName.toUpperCase() === "ARCGIS-VERSION-MANAGEMENT-VERSION-LIST"), b?.append(l), this.arcgisCreateVersionList.emit(l);
    }}></calcite-action>`, g = t`<calcite-tooltip overlay-positioning=fixed .referenceElement=${`action-manage-versions-${i}`}><span>${n.headers.manageVersions}</span></calcite-tooltip>`), t`<calcite-block expanded .heading=${i}>${C}${L}${d}${g}</calcite-block>`;
  }
}
M("arcgis-version-management-service-item", F);
export {
  F as ArcgisVersionManagementServiceItem
};
