// @arcgis/map-components/dist/components/arcgis-link-chart-layout-switcher/customElement.js
/* COPYRIGHT Esri - https://js.arcgis.com/5.0/LICENSE.txt */
import { c as h } from "../../chunks/runtime.js";
import { css as d, html as i } from "lit";
import m from "@arcgis/core/widgets/LinkChartLayoutSwitcher/LinkChartLayoutSwitcherViewModel.js";
import { usePropertyChange as g } from "@arcgis/lumina/controllers";
import { m as p } from "../../chunks/useViewModel.js";
import { u } from "../../chunks/useT9n.js";
import { reEmitEvent as y } from "@arcgis/lumina/controllers/accessor";
import { LitElement as w, createEvent as v } from "@arcgis/lumina";
const L = d``, e = {
  forceDirected: "organic-standard",
  community: "organic-community",
  simple: "basic-grid",
  mechanical: "hierarchical-bottom-to-top",
  radialTree: "radial-root-centric",
  smartTree: "tree-left-to-right",
  geographic: "geographic-organic-standard",
  chronologicalSingle: "chronological-mono-timeline",
  chronologicalMultiple: "chronological-multi-timeline"
}, s = {
  basicLayoutCategory: [e.simple],
  organicLayoutCategory: [
    e.forceDirected,
    e.community,
    e.geographic
  ],
  treeLayoutCategory: [e.smartTree, e.radialTree],
  hierarchicalLayoutCategory: [e.mechanical],
  chronologicalLayoutCategory: [e.chronologicalSingle, e.chronologicalMultiple]
}, l = {
  [e.forceDirected]: "force-directed-link-chart-layout",
  [e.community]: "community-link-chart-layout",
  [e.simple]: "simple-link-chart-layout",
  [e.mechanical]: "hierarchical-link-chart-layout",
  [e.radialTree]: "radial-tree-link-chart-layout",
  [e.smartTree]: "smart-tree-link-chart-layout",
  [e.geographic]: "geographic-link-chart-layout",
  [e.chronologicalSingle]: "single-timeline",
  [e.chronologicalMultiple]: "multi-timeline"
}, n = {
  [e.forceDirected]: "forceDirected",
  [e.community]: "community",
  [e.simple]: "simple",
  [e.mechanical]: "hierarchical",
  [e.radialTree]: "radialTree",
  [e.smartTree]: "smartTree",
  [e.geographic]: "geographic",
  [e.chronologicalSingle]: "chronologicalSingle",
  [e.chronologicalMultiple]: "chronologicalMultiple"
}, M = p(m);
class b extends w {
  constructor() {
    super(...arguments), this.messages = u(), this.viewModel = M(this), this.view = this.viewModel.view, this.appearance = "dropdown", this.autoDestroyDisabled = !1, this.closeOnSelectDisabled = !1, this.icon = "nodes-link", this.layout = this.viewModel.layout, this.preventExtentUpdate = this.viewModel.preventExtentUpdate, this.state = this.viewModel.state, this.arcgisSwitchLayout = y(() => this.viewModel, "switchLayout"), this.arcgisPropertyChange = g()("state"), this.arcgisReady = v();
  }
  static {
    this.properties = { view: 0, appearance: 1, autoDestroyDisabled: 5, closeOnSelectDisabled: 5, icon: 1, label: 1, layout: 32, messageOverrides: 0, preventExtentUpdate: 5, referenceElement: 1, state: 32 };
  }
  static {
    this.styles = L;
  }
  async destroy() {
    await this.manager.destroy();
  }
  async switchLayout(t) {
    this.viewModel.switchLayout(t);
  }
  _renderDropdown() {
    const { messages: t, icon: o } = this;
    return i`<calcite-dropdown scale=s width=m max-items=9 .closeOnSelectDisabled=${this.closeOnSelectDisabled}><calcite-action slot=trigger .disabled=${this.viewModel.state === "disabled"} .icon=${o ?? void 0} .label=${t.layoutOptions ?? "Layout"} scale=s .text=${t.layoutOptions ?? "Layout"}></calcite-action>${Object.entries(s).map(([c, a]) => i`<calcite-dropdown-group selection-mode=single .groupTitle=${this.messages[c]}>${a.map((r) => i`<calcite-dropdown-item .iconStart=${l[r]} .selected=${r === this.layout ? !0 : void 0} @click=${() => void this.switchLayout(r)}>${t[n[r]]}</calcite-dropdown-item>`)}</calcite-dropdown-group>`)}</calcite-dropdown>`;
  }
  _renderMenuBar() {
    const { messages: t } = this;
    return i`<calcite-action-bar scale=s floating overflow-actions-disabled>${Object.entries(s).map(([o, c]) => i`<calcite-action-group .label=${this.messages[o]}>${c.map((a) => i`<calcite-action scale=s .icon=${l[a]} .active=${a === this.layout ? !0 : void 0} @click=${() => void this.switchLayout(a)} .text=${t[n[a]]}></calcite-action>`)}</calcite-action-group>`)}</calcite-action-bar>`;
  }
  render() {
    const { appearance: t } = this, o = t === "menu-bar" ? this._renderMenuBar() : this._renderDropdown();
    return i`<div class="root">${o}</div>`;
  }
}
h("arcgis-link-chart-layout-switcher", b);
export {
  b as ArcgisLinkChartLayoutSwitcher
};
