// @arcgis/map-components/dist/components/arcgis-unit-select/customElement.js
/* COPYRIGHT Esri - https://js.arcgis.com/5.0/LICENSE.txt */
import { c as r } from "../../chunks/runtime.js";
import { repeat as o } from "lit/directives/repeat.js";
import { css as n, html as a } from "lit";
import { LitElement as m, createEvent as u } from "@arcgis/lumina";
import { u as h } from "../../chunks/useT9n.js";
import { i as g } from "../../chunks/measurementUtils.js";
const p = n`:host{width:100%}.root{--calcite-label-margin-bottom: 0;width:100%}.root[layout=inline-space-between] calcite-select{width:var(--arcgis-unit-select-inline-size, 180px)}`;
class y extends m {
  constructor() {
    super(...arguments), this.messages = h({}), this.options = [], this.layout = "block", this.arcgisUnitSelectChange = u();
  }
  static {
    this.properties = { options: [1, { type: Array }], selectLabel: 1, value: 0, messageOverrides: 0, label: 1, layout: 1 };
  }
  static {
    this.styles = p;
  }
  async destroy() {
    await this.manager.destroy();
  }
  render() {
    const {
      messages: t,
      options: i = [],
      // when running in the CI, sometimes this ends up being `undefined`, for unclear reasons.
      selectLabel: s,
      value: l,
      arcgisUnitSelectChange: c
    } = this;
    return a`<calcite-label class="root" .layout=${this.layout}>${s}<calcite-select .label=${s} @calciteSelectChange=${(e) => {
      c.emit(e.target.value);
    }}>${o(i, (e) => e, (e) => a`<calcite-option .selected=${e === l} .value=${e}>${g(e) ? t.systems?.[e] : t.units?.[e]?.pluralCapitalized}</calcite-option>`)}</calcite-select></calcite-label>`;
  }
}
r("arcgis-unit-select", y);
export {
  y as ArcgisUnitSelect
};
