// @arcgis/map-components/dist/components/arcgis-zoom/customElement.js
/* COPYRIGHT Esri - https://js.arcgis.com/5.0/LICENSE.txt */
import { c as u } from "../../chunks/runtime.js";
import { css as d, html as h } from "lit";
import { ignoreAbortErrors as f } from "@arcgis/core/core/promiseUtils.js";
import { when as g } from "@arcgis/core/core/reactiveUtils.js";
import { LitElement as y, createEvent as r, safeClassMap as b, nothing as c } from "@arcgis/lumina";
import { usePropertyChange as p } from "@arcgis/lumina/controllers";
import { u as z } from "../../chunks/useT9n.js";
import { u as v } from "../../chunks/useView.js";
import { g as x } from "../../chunks/globalCss.js";
function n(t) {
  if (!t)
    return !1;
  if (t.type === "3d")
    return t.ready;
  if (!t?.ready)
    return !1;
  const o = m(t), a = t.constraints?.effectiveMaxScale;
  return a === 0 || o > a;
}
function l(t) {
  if (!t)
    return !1;
  if (t.type === "3d")
    return t.ready;
  if (!t?.ready)
    return !1;
  const o = m(t), a = t.constraints?.effectiveMinScale;
  return a === 0 || o < a;
}
function m(t) {
  const e = t?.animation?.target;
  return (e && "then" in e ? void 0 : e?.scale) ?? t?.scale ?? 0;
}
const w = d`.arcgis-button{--calcite-icon-color: var(--calcite-color-text-3);--calcite-button-background-color: var(--calcite-color-foreground-1);color:var(--calcite-color-text-3);line-height:var(--calcite-font-line-height-relative-tight);display:flex;flex-flow:column;cursor:pointer;background-color:var(--calcite-color-foreground-1);font-size:var(--calcite-font-size);:hover{--calcite-icon-color: var(--calcite-color-text-1);--calcite-button-background-color: var(--calcite-color-transparent-hover)}}:host{display:flex;max-width:fit-content}.zoom-out{border-block-start:solid 1px var(--calcite-color-border-1)}.horizontal{flex-flow:row-reverse nowrap}.horizontal .zoom-out{border-block-start:none;border-inline-end:solid 1px var(--calcite-color-border-1)}`;
class $ extends y {
  constructor() {
    super(...arguments), this.messages = z({}), this._state = "disabled", this.view = v(this), this.autoDestroyDisabled = !1, this.icon = "magnifying-glass-plus", this.layout = "vertical", this.visualScale = "m", this.arcgisActionZoom = r(), this.arcgisPropertyChange = p()("state"), this.arcgisReady = r();
  }
  static {
    this.properties = { _state: 16, view: 0, autoDestroyDisabled: 5, icon: 1, label: 1, layout: 3, messageOverrides: 0, referenceElement: 1, visualScale: 1, state: 32 };
  }
  static {
    this.styles = w;
  }
  get state() {
    return this._state;
  }
  async destroy() {
    await this.manager.destroy();
  }
  async zoomIn() {
    await this.zoomBy(1);
  }
  async zoomOut() {
    await this.zoomBy(-1);
  }
  loaded() {
    this.manager.onLifecycle(() => [
      g(() => this.view?.ready ? "ready" : "disabled", (e) => {
        this._state = e;
      }, { sync: !0, initial: !0 })
    ]);
  }
  async zoomBy(e) {
    const { view: o } = this, a = e === 1;
    if (!(!(a ? n(o) : l(o)) || !o))
      if (this.arcgisActionZoom.emit({ action: a ? "zoomIn" : "zoomOut" }), o.type === "2d") {
        const s = o.mapViewNavigation;
        a ? await s.zoomIn() : await s.zoomOut();
      } else
        await f(o.goTo({ zoom: o.zoom + e }));
  }
  render() {
    const e = n(this.view), o = l(this.view), { zoomIn: a, zoomOut: i } = this.messages;
    return h`<div class=${b({
      [x.arcgisButton]: !0,
      horizontal: this.layout === "horizontal"
    })}><calcite-button class="zoom-in" .disabled=${!e} icon-start=plus kind=neutral .label=${a} @click=${this.zoomIn} .scale=${this.visualScale} title=${a ?? c}></calcite-button><calcite-button class="zoom-out" .disabled=${!o} icon-start=minus kind=neutral .label=${i} @click=${this.zoomOut} .scale=${this.visualScale} title=${i ?? c}></calcite-button></div>`;
  }
}
u("arcgis-zoom", $);
export {
  $ as ArcgisZoom
};
