// @arcgis/map-components/dist/components/arcgis-reverse-geocode-list/customElement.js
/* COPYRIGHT Esri - https://js.arcgis.com/5.0/LICENSE.txt */
import { c as u } from "../../chunks/runtime.js";
import { css as m, html as i } from "lit";
import h from "@arcgis/core/portal/Portal.js";
import { LitElement as p, createEvent as g, safeClassMap as b } from "@arcgis/lumina";
import { locationToAddress as f } from "@arcgis/core/rest/locator.js";
import { substitute as _ } from "@arcgis/core/intl.js";
import { u as v } from "../../chunks/useT9n.js";
import { watch as d } from "@arcgis/core/core/reactiveUtils.js";
const $ = m`.coordinates{font-size:var(--calcite-font-size-sm);color:var(--calcite-color-text-3);padding-inline-start:var(--calcite-spacing-md);padding-block-end:var(--calcite-spacing-sm)}`, R = {
  coordinates: "coordinates"
};
class C extends p {
  constructor() {
    super(...arguments), this.messages = v(), this._abortController = null, this._results = null, this.headingLevel = 3, this.portal = h.getDefault(), this.arcgisReverseGeocodeResultsChange = g();
  }
  static {
    this.properties = { _abortController: 16, _results: 16, headingLevel: 9, portal: 0, location: 0, messageOverrides: 0, reverseGeocodeResults: 32 };
  }
  static {
    this.styles = $;
  }
  get reverseGeocodeResults() {
    return this._results;
  }
  async destroy() {
    await this.manager.destroy(), this._cancelQueries();
  }
  loaded() {
    this.manager.onLifecycle(() => [
      d(() => [this.portal, this.location], () => void this._update(), { initial: !0 }),
      d(() => this._results, () => {
        this.arcgisReverseGeocodeResultsChange.emit();
      })
    ]);
  }
  _cancelQueries() {
    this._abortController?.abort(), this._abortController = null;
  }
  async _update() {
    const { location: e, messages: t } = this;
    if (this._cancelQueries(), !e) {
      this._results = null;
      return;
    }
    const s = new AbortController(), r = s.signal;
    this._abortController = s, await this.portal?.load({ signal: r });
    const l = this.portal?.helperServices?.geocode?.filter((o) => o.reverseGeocode ?? !0);
    if (!l) {
      this._results = null;
      return;
    }
    const a = l.map(async (o) => await f(o.url, { location: e }, { signal: r })), n = await Promise.allSettled(a);
    this._abortController = null, this._results = l.map((o, c) => ({
      name: o.name ?? t?.untitledService ?? "",
      address: (n[c].status === "fulfilled" ? n[c].value.address : null) ?? t?.noResultsFound ?? "",
      url: o.url
    }));
  }
  _renderMoreResults(e) {
    const { messages: t } = this;
    return !this._abortController && e?.length ? i`<calcite-block collapsible .headingLevel=${this.headingLevel} .heading=${_(t?.showMoreResults ?? "", { count: e.length })}><calcite-list display-mode=flat interaction-mode=static selection-mode=none .label=${t?.moreResults ?? ""}>${e?.map((s) => i`<calcite-list-item .label=${s.name} .description=${s.address}></calcite-list-item>`)}</calcite-list></calcite-block>` : null;
  }
  render() {
    const { _abortController: e, location: t, _results: s, messages: r } = this;
    if (!t)
      return null;
    const l = s?.slice(), a = l?.shift(), n = i`<div class=${b(R.coordinates)} slot=content-bottom>${`${t.latitude?.toFixed(6)}, ${t.longitude?.toFixed(6)}`}</div>`;
    return i`<calcite-list .loading=${!!e} display-mode=flat interaction-mode=static selection-mode=none .label=${r.label ?? ""}>${a ? i`<calcite-list-item .label=${a.name} .description=${a.address}>${n}</calcite-list-item>` : e ? null : i`<calcite-list-item .label=${r.noResultsFound ?? ""}>${n}</calcite-list-item>`}</calcite-list>${this._renderMoreResults(l)}`;
  }
}
u("arcgis-reverse-geocode-list", C);
export {
  C as ArcgisReverseGeocodeList
};
