// @arcgis/map-components/dist/components/arcgis-feature-media/customElement.js
/* COPYRIGHT Esri - https://js.arcgis.com/5.0/LICENSE.txt */
import { c as O } from "../../chunks/runtime.js";
import { css as B, html as x } from "lit";
import { ref as j } from "lit/directives/ref.js";
import { keyed as W } from "lit/directives/keyed.js";
import { useDirection as q, useMedia as G } from "@arcgis/lumina/controllers";
import { getColorsFromRenderer as K } from "@arcgis/core/renderers/support/utils.js";
import { LitElement as J, nothing as y } from "@arcgis/lumina";
import { substitute as Q } from "@arcgis/core/intl.js";
import { watch as Z } from "@arcgis/core/core/reactiveUtils.js";
import ee from "@arcgis/core/popup/FieldInfo.js";
import N from "@arcgis/core/popup/content/support/ChartMediaInfoValueSeries.js";
import { unsafeGetCalciteModeName as te } from "@arcgis/toolkit/dom";
import { renderingSanitizer as V } from "@arcgis/core/applications/Components/sanitizerUtils.js";
import { d as A, s as R, m as z, n as X, o as ie, p as E, i as ae, q as re, j as Y, f as ne } from "../../chunks/feature-utils.js";
const oe = B`.container{display:flex;flex-flow:row wrap;align-items:center;margin-top:var(--calcite-spacing-xs);width:100%;min-height:150px}.container .pagination-button{display:flex;align-items:center;justify-content:center;border:none;background:transparent;cursor:pointer;padding-block:var(--calcite-spacing-xs);padding-inline:var(--calcite-spacing-xs);color:var(--calcite-color-text-3)}.container .pagination-button:hover,.container .pagination-button:focus{background-color:var(--calcite-color-foreground-3);color:var(--calcite-color-text-1);fill:var(--calcite-color-text-1)}.item-title{margin:0;font-size:var(--calcite-font-size-md);line-height:var(--calcite-font-line-height-relative-snug)}.item-caption{padding-block-start:var(--calcite-spacing-xxs);font-size:var(--calcite-font-size-sm);line-height:var(--calcite-font-line-height-relative-snug)}.item{display:flex;align-items:flex-start;justify-content:center;margin-block:var(--calcite-spacing-xs);width:100%;height:auto}.item-navigation{display:flex;flex:0 1 100%;align-items:flex-start;justify-content:space-between;margin-block-end:var(--calcite-spacing-xs)}.item-text{margin-inline-end:var(--calcite-spacing-xs)}.pagination{display:flex;flex:0 0 auto;align-items:center;overflow:hidden}.pagination:only-child{margin-inline-start:auto}.pagination-text{padding-inline:var(--calcite-spacing-xxs);color:var(--calcite-color-text-3);font-size:var(--calcite-font-size-sm);line-height:var(--calcite-font-line-height-relative-normal)}.item-container{flex:0 1 auto;width:100%}.item-container img{max-width:100%}.item-container img[src$=".SVG"],.item-container img[src$=".svg"]{width:100%}.chart{width:100%;height:155px}`;
function $(S) {
  if (!A(S))
    return null;
  const [e, t] = S.split("/").slice(1);
  return { layerId: e, fieldName: t };
}
const k = 15, M = "category", w = "value", se = "rgba(50, 50, 50, 1)", le = 250, ce = 500, de = 200;
class ue extends J {
  constructor() {
    super(...arguments), this._direction = q(), this._prefersReducedMotion = G("(prefers-reduced-motion)"), this.activeMediaInfoIndex = 0, this.chartAnimationDisabled = !1, this.headingLevel = 2;
  }
  static {
    this.properties = { _refreshIntervalInfo: 16, messages: 0, values: 0, activeMediaInfoIndex: 9, chartAnimationDisabled: 5, activeMediaInfo: 32, formattedMediaInfos: 32, description: 1, fieldInfoMap: 0, layer: 0, mediaInfos: 0, popupTemplate: 0, relatedInfos: 0, formattedValues: 0, heading: 1, graphic: 0, formattedMediaInfoCount: 32, headingLevel: 9 };
  }
  static {
    this.styles = oe;
  }
  get activeMediaInfo() {
    return this.formattedMediaInfos[this.activeMediaInfoIndex] || null;
  }
  get formattedMediaInfos() {
    return this._formatMediaInfos() || [];
  }
  get formattedMediaInfoCount() {
    return this.formattedMediaInfos.length;
  }
  next() {
    this._pageContentElementMedia(1);
  }
  previous() {
    this._pageContentElementMedia(-1);
  }
  setActiveMedia(e) {
    this._setContentElementMedia(e);
  }
  loaded() {
    this.manager.onLifecycle(() => [
      Z(() => [this.activeMediaInfo, this.activeMediaInfoIndex], () => this._setupMediaRefreshTimer(), { initial: !0 }),
      { remove: () => this._clearMediaRefreshTimer() }
    ]);
  }
  _setContentElementMedia(e) {
    const { formattedMediaInfoCount: t } = this, i = (e + t) % t;
    this.activeMediaInfoIndex = i;
  }
  _pageContentElementMedia(e) {
    const { activeMediaInfoIndex: t } = this, i = t + e;
    this._setContentElementMedia(i);
  }
  _formatMediaInfos() {
    const { mediaInfos: e, layer: t } = this, i = this.values ?? {}, a = this.formattedValues ?? {}, r = this.fieldInfoMap ?? /* @__PURE__ */ new Map();
    return e?.map((o) => {
      const n = o?.clone();
      if (!n)
        return null;
      if (n.title = R({
        values: i,
        fieldInfoMap: r,
        globalValues: a,
        layer: t,
        text: n.title
      }), n.caption = R({
        values: i,
        fieldInfoMap: r,
        globalValues: a,
        layer: t,
        text: n.caption
      }), n.altText = R({
        values: i,
        fieldInfoMap: r,
        globalValues: a,
        layer: t,
        text: n.altText
      }), n.type === "image") {
        if (!n.value)
          return null;
        const s = n.value;
        return this._setImageValue({
          value: s,
          formattedValues: a,
          layer: t
        }), s.sourceURL ? n : void 0;
      }
      if (n.type === "pie-chart" || n.type === "line-chart" || n.type === "column-chart" || n.type === "bar-chart") {
        if (!n.value)
          return null;
        const s = n.value;
        return this._setChartValue({
          value: s,
          chartType: n.type,
          values: i,
          formattedValues: a,
          layer: t
        }), n;
      }
      return null;
    }).filter((o) => o != null) ?? [];
  }
  _setImageValue(e) {
    const t = this.fieldInfoMap ?? /* @__PURE__ */ new Map(), { value: i, formattedValues: a, layer: r } = e, { linkURL: o, sourceURL: n } = i;
    if (n) {
      const s = z(n, r);
      i.sourceURL = X({
        formattedValues: a,
        template: s,
        fieldInfoMap: t
      });
    }
    if (o) {
      const s = z(o, r);
      i.linkURL = X({
        formattedValues: a,
        template: s,
        fieldInfoMap: t
      });
    }
  }
  _setChartValue(e) {
    const { value: t, values: i, formattedValues: a, chartType: r, layer: o } = e, { popupTemplate: n, relatedInfos: s } = this, { fields: c, normalizeField: v } = t, d = o;
    if (t.fields = ie(c, d), v && (t.normalizeField = E(v, d)), !c.some((u) => !!(a[u] != null || A(u) && s?.size)))
      return;
    const h = n?.fieldInfos ?? [];
    t.series = [], c.forEach((u, f) => {
      const m = t.colors?.[f];
      if (A(u)) {
        t.series = [
          ...t.series,
          ...this._getRelatedChartInfos({ fieldInfos: h, fieldName: u, formattedValues: a, chartType: r, value: t, color: m })
        ];
        return;
      }
      const p = this._getChartOption({
        value: t,
        values: i,
        chartType: r,
        formattedValues: a,
        fieldName: u,
        fieldInfos: h,
        color: m
      });
      t.series.push(p);
    });
  }
  _getRelatedChartInfos(e) {
    const { fieldInfos: t, fieldName: i, formattedValues: a, chartType: r, value: o, color: n } = e, s = [], c = $(i), v = c && this.relatedInfos?.get(c.layerId);
    if (!v)
      return s;
    const { relatedFeatures: d, relation: l } = v;
    if (!l || !d)
      return s;
    const { cardinality: h } = l;
    return d.forEach((f) => {
      const m = f.attributes;
      m && Object.keys(m).forEach((p) => {
        p === c.fieldName && s.push(this._getChartOption({
          value: o,
          values: m,
          formattedValues: a,
          fieldName: i,
          chartType: r,
          relatedFieldName: p,
          hasMultipleRelatedFeatures: d?.length > 1,
          fieldInfos: t,
          color: n
        }));
      });
    }), h && (h === "one-to-many" || h === "many-to-many") ? s : [s[0]];
  }
  _getTooltip({ label: e, value: t, chartType: i }) {
    return i === "pie-chart" ? `${e}` : `${e}: ${t}`;
  }
  _getChartOption(e) {
    const { value: t, values: i, formattedValues: a, fieldName: r, relatedFieldName: o, fieldInfos: n, chartType: s, hasMultipleRelatedFeatures: c, color: v } = e, { layer: d, graphic: l } = this, h = this.fieldInfoMap ?? /* @__PURE__ */ new Map(), { normalizeField: u, tooltipField: f } = t, m = u ? A(u) ? i[$(u).fieldName] : i[u] : null, p = ae(r) && a?.[r] !== void 0 ? a[r] : o && i[o] !== void 0 ? i[o] : i[r] !== void 0 ? i[r] : a[r], I = p === void 0 ? null : p && m && typeof p == "number" && typeof m == "number" ? p / m : p, _ = re(n, r), g = E(r, d), C = (f && a[f] !== void 0 ? a[f] : Y({
      fieldInfo: _ || new ee({ fieldName: g }),
      expressionInfos: this.popupTemplate?.expressionInfos,
      graphic: l,
      layer: d
    })) ?? "", b = a[g] ?? "", L = I != null ? typeof I == "number" ? I : Number(I) : null;
    if (A(r)) {
      const T = h.get(r.toLowerCase()), P = f ? h.get(f.toLowerCase()) : null, U = T?.fieldName ?? r, F = c && f ? $(f).fieldName : P?.fieldName ?? f, H = (c && F ? i[F] : a[F] ?? (T && Y({ fieldInfo: T, layer: d, graphic: l })) ?? T?.fieldName ?? o) ?? "", D = (c && o ? i[o] : a[U]) ?? "";
      return new N({
        fieldName: r,
        tooltip: this._getTooltip({
          label: H,
          value: D,
          chartType: s
        }),
        color: v,
        value: L
      });
    }
    return new N({
      fieldName: r,
      tooltip: this._getTooltip({
        label: C,
        value: b,
        chartType: s
      }),
      color: v,
      value: L
    });
  }
  _getRenderer() {
    const e = this.layer;
    return this.graphic?.isAggregate && e?.featureReduction && "renderer" in e.featureReduction ? e.featureReduction.renderer : e?.renderer;
  }
  async _getSeriesColors(e) {
    const { color: t } = await import("@amcharts/amcharts5/index.js"), i = /* @__PURE__ */ new Map();
    return e.forEach((a) => {
      a.color && i.set(a, t(a.color.toCss(!0)));
    }), i;
  }
  async _getRendererColors() {
    const { color: e } = await import("@amcharts/amcharts5/index.js"), t = /* @__PURE__ */ new Map(), i = this._getRenderer(), a = "default";
    if (!i)
      return t;
    const r = await K(i);
    return r.delete(a), Array.from(r.values()).every((n) => n?.length === 1) && Array.from(r.keys()).forEach((n) => {
      const s = r.get(n)?.[0]?.toCss(!0);
      s && t.set(n, e(s));
    }), t;
  }
  async _createChart(e) {
    const { activeMediaInfo: t } = this;
    if (!e || !t)
      return;
    const { createRoot: i } = await import("../../chunks/chart-utils.js"), a = await i(e);
    this._chartRoot = a, this._renderChart({
      mediaInfo: t,
      root: a
    });
  }
  _handleMediaKeyup(e) {
    const { key: t } = e;
    t === "ArrowLeft" && (e.stopPropagation(), this.previous()), t === "ArrowRight" && (e.stopPropagation(), this.next());
  }
  _canAnimateChart() {
    return !this.chartAnimationDisabled && !this._prefersReducedMotion;
  }
  _getChartAnimationMS() {
    return this._canAnimateChart() ? le : 0;
  }
  _getChartSeriesAnimationMS() {
    return this._canAnimateChart() ? ce : 0;
  }
  async _customizeChartTooltip(e, t = "horizontal") {
    const { color: i } = await import("@amcharts/amcharts5/index.js");
    e.setAll({
      pointerOrientation: t
    }), e.get("background")?.setAll({
      stroke: i(se)
    }), e.label.setAll({
      direction: this._direction,
      oversizedBehavior: "wrap",
      maxWidth: de
    });
  }
  async _createPieChart(e, t) {
    const [{ Tooltip: i }, { PieChart: a, PieSeries: r }] = await Promise.all([
      import("@amcharts/amcharts5/index.js"),
      import("@amcharts/amcharts5/percent.js")
    ]), { mediaInfo: o, root: n } = e, { title: s } = o, c = 5, v = o?.altText || o?.title || "", d = n.container.children.push(a.new(n, {
      role: "figure",
      ariaLabel: v,
      focusable: !0,
      paddingBottom: c,
      paddingTop: c,
      paddingLeft: c,
      paddingRight: c
    })), l = `{category}: {valuePercentTotal.formatNumber('0.00')}%
 ({value})`, h = i.new(n, {
      labelText: l
    }), u = d.series.push(r.new(n, {
      name: s,
      valueField: w,
      categoryField: M,
      tooltip: h
    }));
    u.ticks.template.set("forceHidden", !0), u.labels.template.set("forceHidden", !0), u.slices.template.states.create("active", {
      shiftRadius: c
    }), this._customizeChartTooltip(h), u.slices.template.setAll({
      ariaLabel: l,
      focusable: !0,
      templateField: "columnSettings"
    }), u.data.setAll(t), u.appear(this._getChartSeriesAnimationMS()), d.appear(this._getChartAnimationMS()), d.root.dom.classList.toggle("chart-rendered", !0);
  }
  _getMinSeriesValue(e) {
    let t = 0;
    return e.forEach((i) => t = Math.min(i.value, t)), t;
  }
  async _createColumnChart(e, t, i) {
    const [{ Tooltip: a, Scrollbar: r }, { CategoryAxis: o, AxisRendererX: n, AxisRendererY: s, ValueAxis: c, ColumnSeries: v }] = await Promise.all([import("@amcharts/amcharts5/index.js"), import("@amcharts/amcharts5/xy.js")]), { mediaInfo: d, root: l } = t, { value: h, title: u } = d, { _direction: f } = this;
    e.setAll({
      wheelX: "panX",
      wheelY: "zoomX"
    });
    const m = e.xAxes.push(o.new(l, {
      renderer: n.new(l, {
        inversed: f === "rtl"
      }),
      categoryField: M
    }));
    m.get("renderer").labels.template.setAll({
      forceHidden: !0
    });
    const p = e.yAxes.push(c.new(l, {
      renderer: s.new(l, { inside: !1 }),
      min: this._getMinSeriesValue(h.series)
    }));
    p.get("renderer").labels.template.setAll({
      direction: f
    });
    const I = "{categoryX}", _ = a.new(l, {
      labelText: I
    }), g = e.series.push(v.new(l, {
      name: u,
      xAxis: m,
      yAxis: p,
      valueYField: w,
      categoryXField: M,
      tooltip: _
    }));
    this._customizeChartTooltip(_), g.columns.template.setAll({
      ariaLabel: I,
      focusable: !0,
      templateField: "columnSettings"
    }), h.series.length > k && e.set("scrollbarX", r.new(l, {
      orientation: "horizontal"
    })), m.data.setAll(i), g.data.setAll(i), g.appear(this._getChartSeriesAnimationMS()), e.appear(this._getChartAnimationMS());
  }
  async _createBarChart(e, t, i) {
    const [{ Tooltip: a, Scrollbar: r }, { CategoryAxis: o, AxisRendererX: n, AxisRendererY: s, ValueAxis: c, ColumnSeries: v }] = await Promise.all([import("@amcharts/amcharts5/index.js"), import("@amcharts/amcharts5/xy.js")]), { mediaInfo: d, root: l } = t, { value: h, title: u } = d, { _direction: f } = this;
    e.setAll({
      wheelX: "panY",
      wheelY: "zoomY"
    });
    const m = e.yAxes.push(o.new(l, {
      renderer: s.new(l, {
        inversed: !0
      }),
      categoryField: M
    }));
    m.get("renderer").labels.template.setAll({
      forceHidden: !0
    });
    const p = e.xAxes.push(c.new(l, {
      renderer: n.new(l, {
        inside: !1,
        inversed: f === "rtl"
      }),
      min: this._getMinSeriesValue(h.series)
    }));
    p.get("renderer").labels.template.setAll({
      direction: f
    });
    const I = "{categoryY}", _ = a.new(l, {
      labelText: I
    }), g = e.series.push(v.new(l, {
      name: u,
      xAxis: p,
      yAxis: m,
      valueXField: w,
      categoryYField: M,
      tooltip: _
    }));
    this._customizeChartTooltip(_, "vertical"), g.columns.template.setAll({
      ariaLabel: I,
      focusable: !0,
      templateField: "columnSettings"
    }), h.series.length > k && e.set("scrollbarY", r.new(l, {
      orientation: "vertical"
    })), m.data.setAll(i), g.data.setAll(i), g.appear(this._getChartSeriesAnimationMS()), e.appear(this._getChartAnimationMS());
  }
  async _createLineChart(e, t, i) {
    const [{ Tooltip: a, Scrollbar: r }, { CategoryAxis: o, AxisRendererX: n, AxisRendererY: s, ValueAxis: c, LineSeries: v }] = await Promise.all([import("@amcharts/amcharts5/index.js"), import("@amcharts/amcharts5/xy.js")]), { root: d, mediaInfo: l } = t, { value: h, title: u } = l, { _direction: f } = this;
    e.setAll({
      wheelX: "panX",
      wheelY: "zoomX"
    });
    const m = e.xAxes.push(o.new(d, {
      renderer: n.new(d, {
        inversed: f === "rtl"
      }),
      categoryField: M
    }));
    m.get("renderer").labels.template.setAll({
      forceHidden: !0
    });
    const p = e.yAxes.push(c.new(d, {
      renderer: s.new(d, { inside: !1 }),
      min: this._getMinSeriesValue(h.series)
    }));
    p.get("renderer").labels.template.setAll({
      direction: f
    });
    const I = "{categoryX}", _ = i[0]?.lineSettings?.stroke, g = a.new(d, {
      getFillFromSprite: !_,
      labelText: I
    });
    _ && g.get("background")?.setAll({
      fill: _
    });
    const C = e.series.push(v.new(d, {
      name: u,
      xAxis: m,
      yAxis: p,
      valueYField: w,
      categoryXField: M,
      tooltip: g
    }));
    C.strokes.template.setAll({
      templateField: "lineSettings"
    }), this._customizeChartTooltip(g, "vertical"), h.series.length > k && e.set("scrollbarX", r.new(d, {
      orientation: "horizontal"
    })), m.data.setAll(i), C.data.setAll(i), C.appear(this._getChartSeriesAnimationMS()), e.appear(this._getChartAnimationMS());
  }
  async _createXYChart(e, t) {
    const { XYChart: i, XYCursor: a } = await import("@amcharts/amcharts5/xy.js"), { root: r, mediaInfo: o } = e, { type: n } = o, s = o?.altText || o?.title || "", c = r.container.children.push(i.new(r, {
      role: "figure",
      ariaLabel: s,
      focusable: !0,
      panX: !0,
      panY: !0
    }));
    c.set("cursor", a.new(r, {})), n === "column-chart" && await this._createColumnChart(c, e, t), n === "bar-chart" && await this._createBarChart(c, e, t), n === "line-chart" && await this._createLineChart(c, e, t), c.root.dom.classList.toggle("chart-rendered", !0);
  }
  _clearMediaRefreshTimer() {
    const { _refreshTimer: e } = this;
    e && (clearTimeout(e), this._refreshTimer = void 0);
  }
  _updateMediaInfoTimestamp(e) {
    const t = Date.now();
    this._refreshIntervalInfo = {
      timestamp: t,
      sourceURL: e && this._getImageSource(e, t)
    };
  }
  _setupMediaRefreshTimer() {
    this._clearMediaRefreshTimer();
    const { activeMediaInfo: e } = this;
    e?.type === "image" && e.refreshInterval > 0 && this._setRefreshTimeout(e);
  }
  _setRefreshTimeout(e) {
    const { refreshInterval: t, value: i } = e, a = t * 6e4;
    this._updateMediaInfoTimestamp(i.sourceURL);
    const r = window.setInterval(() => {
      this._updateMediaInfoTimestamp(i.sourceURL);
    }, a);
    this._refreshTimer = r;
  }
  _getImageSource(e, t) {
    const i = e.includes("?") ? "&" : "?", [a, r = ""] = e.split("#");
    return `${a}${i}timestamp=${t}${r ? "#" : ""}${r}`;
  }
  _updateChartElement(e) {
    this._chartRoot?.dispose(), this._chartRoot = void 0, e && this._createChart(e);
  }
  render() {
    return x`<div class="root" @keyup=${this._handleMediaKeyup}><arcgis-feature-element-info .heading=${this.heading} .description=${this.description} .headingLevel=${this.headingLevel}></arcgis-feature-element-info>${this._renderMedia()}</div>`;
  }
  _renderMedia() {
    const { formattedMediaInfoCount: e, activeMediaInfoIndex: t } = this, i = this._renderMediaText();
    return e ? x`<div class="container">${this._renderMediaInfo()}<div class="item-navigation">${i}${e > 1 ? x`<div class="pagination">${this._renderMediaPageButton("previous")}<span class="pagination-text">${Q(this.messages?.pageText ?? "", {
      index: t + 1,
      total: e
    })}</span>${this._renderMediaPageButton("next")}</div>` : null}</div></div>` : null;
  }
  _renderMediaText() {
    const { activeMediaInfo: e } = this;
    if (!e)
      return null;
    const t = e?.title ? x`<div class="item-title" .innerHTML=${V.sanitize(e.title) ?? ""}></div>` : null, i = e?.caption ? x`<div class="item-caption" .innerHTML=${V.sanitize(e.caption) ?? ""}></div>` : null;
    return t || i ? x`<div class="item-text">${t}${i}</div>` : null;
  }
  _renderImageMediaInfo(e) {
    if (!e.value)
      return null;
    const { _refreshIntervalInfo: t } = this, { value: i, refreshInterval: a, altText: r, title: o } = e, { sourceURL: n, linkURL: s } = i, v = ne(s ?? void 0) ? "_blank" : "_self", d = v === "_blank" ? "noreferrer" : "", l = a ? t : null, h = l ? l.sourceURL : n, u = x`<img alt=${(r || o) ?? y} src=${h ?? void 0 ?? y}>`;
    return (s ? x`<a href=${s ?? y} rel=${d ?? y} target=${v} title=${o ?? y}>${u}</a>` : null) ?? u;
  }
  _renderChartMediaInfo(e) {
    const { activeMediaInfoIndex: t, formattedMediaInfoCount: i } = this;
    return W(`media-${e.type}-${t}-${i}`, x`<div class="chart" ${j(this._updateChartElement)}></div>`);
  }
  _renderMediaInfoType() {
    const { activeMediaInfo: e } = this;
    return e ? e.type === "image" ? this._renderImageMediaInfo(e) : e.type.includes("chart") ? this._renderChartMediaInfo(e) : null : null;
  }
  _renderMediaInfo() {
    const { activeMediaInfo: e } = this;
    return e ? x`<div class="item-container"><div class="item">${this._renderMediaInfoType()}</div></div>` : null;
  }
  _renderMediaPageButton(e) {
    if (this.formattedMediaInfoCount < 2)
      return null;
    const t = e === "previous", i = t ? this.messages?.previous ?? "" : this.messages?.next ?? "", a = t ? "chevron-left" : "chevron-right", r = t ? this.previous : this.next;
    return x`<button .ariaLabel=${i} class="pagination-button" @click=${r} tabindex=0 title=${i ?? y} type=button><calcite-icon flip-rtl class="pagination-icon" .icon=${a} scale=s></calcite-icon></button>`;
  }
  async _renderChart(e) {
    const { root: t, mediaInfo: i } = e, { value: a, type: r } = i, [{ default: o }, { default: n }, { default: s }, { esriChartColorSet: c }, { ColorSet: v, Theme: d }] = await Promise.all([
      import("@amcharts/amcharts5/themes/Animated.js"),
      import("@amcharts/amcharts5/themes/Dark.js"),
      import("@amcharts/amcharts5/themes/Responsive.js"),
      import("../../chunks/chartCommon.js"),
      import("@amcharts/amcharts5/index.js")
    ]), l = d.new(t);
    l.rule("ColorSet").set("colors", c), l.rule("ColorSet").set("reuse", !0);
    const h = [s.new(t), l];
    te(this.el) === "dark" && h.push(n.new(t)), this._canAnimateChart() && h.push(o.new(t)), t.setThemes(h);
    const u = await this._getRendererColors(), f = await this._getSeriesColors(a.series), m = v.new(t, {}), p = f.get(a.series[0]), I = p ? {
      lineSettings: {
        stroke: p
      }
    } : void 0, _ = a.series.map((g, C) => {
      const b = f.get(g) || u.get(g.fieldName) || m.getIndex(C);
      return {
        [M]: g.tooltip,
        [w]: g.value,
        columnSettings: {
          fill: b,
          stroke: b
        },
        ...I
      };
    }).filter((g) => r === "pie-chart" ? g.value != null && g.value > 0 : !0);
    r === "pie-chart" ? this._createPieChart(e, _) : this._createXYChart(e, _);
  }
}
O("arcgis-feature-media", ue);
export {
  ue as ArcgisFeatureMedia
};
