// @arcgis/map-components/dist/components/arcgis-video/customElement.js
/* COPYRIGHT Esri - https://js.arcgis.com/5.0/LICENSE.txt */
import { c as o } from "../../chunks/runtime.js";
import { css as h, html as e } from "lit";
import { useDirection as n } from "@arcgis/lumina/controllers";
import { watch as t } from "@arcgis/core/core/reactiveUtils.js";
import l from "@arcgis/core/views/VideoView.js";
import { LitElement as c, createEvent as m, safeClassMap as d } from "@arcgis/lumina";
import { makeAccessorController as y } from "@arcgis/lumina/controllers/accessor";
import { createRef as s, ref as r } from "lit/directives/ref.js";
import { s as f, C as a, S as p } from "../../chunks/slots.js";
import { s as w } from "../../chunks/esri-main.js";
const v = h`:host{width:100%;height:100%}`, u = y(l, {
  omitProps: ["container", "navigation", "scale", "ui"]
});
class g extends c {
  constructor() {
    super(), this._containerRef = s(), this._direction = n(), this._mapRef = s(), this.autoDestroyDisabled = !1, this.view = u(this), this.attributionMode = this.view.attributionMode, this.ready = this.view.ready, this.layer = this.view.layer, this.map = this.view.map, this.arcgisViewReadyChange = m();
  }
  static {
    this.properties = { autoDestroyDisabled: 5, view: 32, aria: 0, attributionMode: 1, ready: 36, layer: 0, map: 0 };
  }
  static {
    this.styles = [w, f, v];
  }
  get aria() {
    return this.view?.aria;
  }
  set aria(i) {
    this.view && (this.view.aria = i);
  }
  async destroy() {
    await this.manager.destroy();
  }
  loaded() {
    const i = this;
    i.el.childElem = this._mapRef.value, this.view.container ??= i.el.childElem, this.manager.onLifecycle(() => [
      t(() => this.layer, () => {
        this.layer && this.view && (this.view.layer = this.layer, this.layer.loaded || this.layer.load());
      }, { initial: !0 }),
      t(() => this.view.ready, () => {
        this.view.ready && this.arcgisViewReadyChange.emit();
      }, { initial: this.view.ready })
    ]);
  }
  renderMain() {
    return e`<div class=${`root ${a.base}`} ${r(this._mapRef)}></div>`;
  }
  render() {
    return e`${this.renderMain()}<section class=${d(a.containerUI)} ${r(this._containerRef)}>${p({ direction: this._direction })}</section>`;
  }
}
o("arcgis-video", g);
export {
  g as ArcgisVideo
};
