// @arcgis/map-components/dist/components/arcgis-legend-classic-view/customElement.js
/* COPYRIGHT Esri - https://js.arcgis.com/5.0/LICENSE.txt */
import { c as n } from "../../chunks/runtime.js";
import { css as c, html as t } from "lit";
import { LitElement as d, safeClassMap as o } from "@arcgis/lumina";
import { classes as g } from "@arcgis/toolkit/dom";
import { H as m } from "../../chunks/heading.js";
const h = c`.esri-widget__heading{color:var(--calcite-color-text-2);margin:0 0 .5rem;font-weight:var(--calcite-font-weight-semibold)}h1.esri-widget__heading{font-size:var(--calcite-font-size-xl)}h2.esri-widget__heading{font-size:var(--calcite-font-size-md)}h3.esri-widget__heading,h4.esri-widget__heading,h5.esri-widget__heading,h6.esri-widget__heading{font-size:var(--calcite-font-size)}:host{display:block}.no-legend{padding:var(--calcite-spacing-sm) var(--calcite-spacing-md)}.layer{border-bottom:1px solid var(--calcite-color-border-2);padding:var(--arcgis-internal-padding, var(--calcite-spacing-md) var(--calcite-spacing-lg));word-wrap:break-word}.layer:last-child{border-bottom:none}.layer-elements{margin-left:var(--calcite-spacing-sm);overflow:auto}.layer-child{margin-left:var(--calcite-spacing-sm);padding-right:0;padding-left:0}:dir(rtl) .layer-elements,:dir(rtl) .layer-child{margin-right:var(--calcite-spacing-sm);margin-left:0;overflow:auto}`;
class p extends d {
  constructor() {
    super(...arguments), this.headingLevel = 3, this.type = "classic";
  }
  static {
    this.properties = { activeLayerInfos: 0, headingLevel: 9, type: 32, messages: 0 };
  }
  static {
    this.styles = h;
  }
  render() {
    const s = this.activeLayerInfos?.toArray().map((r) => this._renderLegendForLayer(r)).filter(Boolean);
    return s?.length ? t`<section>${s}</section>` : t`<div class="no-legend">${this.messages.noLegend}</div>`;
  }
  _renderLegendForLayer(e) {
    if (!e.ready)
      return null;
    const s = e.title ? m({
      level: this.headingLevel,
      children: e.title
    }) : null;
    if (e.children.length) {
      const r = e.children.toArray().map((i) => this._renderLegendForLayer(i));
      return t`<section class="layer">${s}${r}</section>`;
    }
    return this._renderLegendElementsForLayer(e, s);
  }
  _renderLegendElementsForLayer(e, s) {
    const r = e.legendElements?.filter(Boolean);
    if (!r?.length)
      return null;
    const i = r.map((a) => t`<arcgis-legend-classic-element .cssEffectFilter=${e.cssEffectFilter ?? void 0} .layer=${e.layer} .legendElement=${a} .messages=${this.messages}></arcgis-legend-classic-element>`), l = { "layer-child": !!e.parent };
    return t`<section class=${o(g("layer", l))}>${s}<div class="layer-elements">${i}</div></section>`;
  }
}
n("arcgis-legend-classic-view", p);
export {
  p as LegendClassicView
};
