// @arcgis/map-components/dist/components/arcgis-video-player-action-bar/customElement.js
/* COPYRIGHT Esri - https://js.arcgis.com/5.0/LICENSE.txt */
import { c as a } from "../../chunks/runtime.js";
import { css as r, html as e } from "lit";
import { formatDate as l } from "@arcgis/core/intl.js";
import { LitElement as c, safeClassMap as n } from "@arcgis/lumina";
import { watch as d } from "@arcgis/core/core/reactiveUtils.js";
import { classes as m } from "@arcgis/toolkit/dom";
const v = r`.toolbar{display:flex;align-items:center;background-color:var(--calcite-color-foreground-1);font-size:var(--calcite-font-size--2)}.timecode{color:var(--calcite-color-text-3);line-height:1.25rem}.controls-group{display:flex;flex:1 0 auto;justify-content:flex-end}.live-icon{color:var(--calcite-color-status-danger)}.controls-container{display:flex;align-items:center;gap:var(--calcite-spacing-md)}`;
function t(s) {
  return l(Math.round(s) * 1e3, { minute: "2-digit", second: "2-digit" });
}
class g extends c {
  static {
    this.properties = { _isLive: 16, viewModel: 0, messages: 0, currentTime: 9, duration: 9, toggleMetadata: 0, inline: 5 };
  }
  static {
    this.styles = v;
  }
  loaded() {
    this.manager.onLifecycle(() => [
      d(() => this.viewModel.layer?.isLive, (i) => {
        this._isLive = i ?? void 0;
      }, { initial: !0 })
    ]);
  }
  render() {
    const { messages: i } = this;
    this._isLive === void 0 && (this._isLive = this.viewModel.layer?.isLive);
    const o = this._isLive ? t(this.currentTime) : `${t(this.currentTime)} / ${t(this.duration)}`;
    return e`<div class=${n(m("toolbar"))}><div class="controls-container"><arcgis-video-player-actions-group .inline=${this.inline} .messages=${i} .viewModel=${this.viewModel}></arcgis-video-player-actions-group>${this._isLive ? e`<calcite-icon class="live-icon" icon=beacon></calcite-icon>` : null}${!this.inline || this._isLive ? e`<div class="timecode">${o}</div>` : null}</div>${this.inline ? null : e`<arcgis-video-player-controls-group class="controls-group" .messages=${i} .toggleMetadata=${() => this.toggleMetadata()} .viewModel=${this.viewModel}></arcgis-video-player-controls-group>`}</div>`;
  }
}
a("arcgis-video-player-action-bar", g);
export {
  g as VideoPlayerActionBar
};
