// @arcgis/map-components/dist/components/arcgis-time-of-day-slider/customElement.js
/* COPYRIGHT Esri - https://js.arcgis.com/5.0/LICENSE.txt */
import { c as u } from "../../chunks/runtime.js";
import { keyed as k } from "lit/directives/keyed.js";
import { substitute as T } from "@arcgis/core/intl.js";
import { LitElement as v, createEvent as f, safeClassMap as _ } from "@arcgis/lumina";
import { css as F, html as d } from "lit";
import { u as L } from "../../chunks/useT9n.js";
import { c as i } from "../../chunks/time-utils.js";
const w = F`:host{display:block}arcgis-slider{--arcgis-slider-handle-size: .875rem;--ticks-font-size: .625rem;--label-font-size: .75rem;--label-padding-block-start: 3px;--label-padding-block-end: var(--calcite-spacing-lg);--calcite-font-size-sm: var(--label-font-size);display:block;margin-block-start:var(--padding-block-start, -.5rem);padding-block-end:var(--padding-block-end, .3rem);&.wrap-labels{padding-block-end:var(--wrapped-padding, .875rem)}}:host([visual-scale="s"]) arcgis-slider{--arcgis-slider-handle-size: .625rem;--ticks-font-size: .625rem;--label-font-size: .625rem;--label-padding-block-end: 1rem;--padding-block-end: .5rem;--wrapped-padding: 1.1rem}:host([visual-scale="l"]) arcgis-slider{--arcgis-slider-handle-size: 1rem;--ticks-font-size: .75rem;--label-font-size: .875rem;--label-padding-block-end: .94rem;--padding-block-end: .4rem;--wrapped-padding: 1.1rem}.ticks-container{width:100%;position:relative}.time-range-label{position:relative;width:100%;padding-block-start:var(--label-padding-block-start);padding-block-end:var(--label-padding-block-end);font-size:var(--label-font-size);font-variant-numeric:tabular-nums;line-height:1em;text-align:center}.main-ticks,.secondary-ticks{position:absolute;inset:0}.main-ticks{--arcgis-ticks-label-width: min-content;--arcgis-ticks-label-text-align: center;--arcgis-ticks-font-size: var(--ticks-font-size);white-space:nowrap;word-break:auto-phrase;font-variant-numeric:tabular-nums}.wrap-labels .main-ticks{white-space:normal}.secondary-ticks{--arcgis-ticks-tick-length: 3px;--arcgis-ticks-gap: 16px}`;
class z extends v {
  constructor() {
    super(...arguments), this.messages = L({ blocking: !0 }), this._shouldWrapLabels = !1, this._sliderLabelFormatter = (e) => "", this._tickLabelFormatter = (e) => "", this.disabled = !1, this.mode = "range", this.startTime = 0, this.endTime = 0, this.visualScale = "m", this.arcgisTimeChange = f();
  }
  static {
    this.properties = { _shouldWrapLabels: 16, _values: 16, _sliderLabelFormatter: 16, _tickLabelFormatter: 16, disabled: 7, mode: 3, startTime: 9, endTime: 9, label: 1, steps: 9, visualScale: 3 };
  }
  static {
    this.styles = w;
  }
  willUpdate(e) {
    (e.has("mode") || e.has("startTime") || e.has("endTime")) && (this._values = this.mode === "single" ? [this.startTime] : [this.startTime, this.endTime]), e.has("messages") && this._updateLabelFormatters();
  }
  _getSliderLabelFormatter(e) {
    const t = new Intl.DateTimeFormat(e, h);
    return (a) => {
      const n = i(a, "minutes", "milliseconds"), m = new Date(n);
      let r = t.format(m);
      return n !== 0 && r === "00:00" && (r = "24:00"), r;
    };
  }
  _getTickLabelFormatter(e) {
    const t = new Intl.DateTimeFormat(e, h), a = new Intl.DateTimeFormat(e, y);
    return (n) => {
      const m = Math.round(i(n, "minutes", "hours")), r = i(m, "hours", "milliseconds"), c = new Date(r);
      let s = t.format(c);
      if (s.includes(" ")) {
        s = a.format(c), s = s.replace(":00", "");
        const b = "⁠";
        s = s.split(" ").map((p) => Array.from(p).join(b)).join(" ");
      }
      return s;
    };
  }
  _setStartTime(e) {
    this.startTime = e, this.arcgisTimeChange.emit();
  }
  _setRange(e, t) {
    this.startTime = e, this.endTime = t, this.arcgisTimeChange.emit();
  }
  _updateLabelFormatters() {
    const e = this.messages._lang;
    this._sliderLabelFormatter = this._getSliderLabelFormatter(e), this._tickLabelFormatter = this._getTickLabelFormatter(e), this._shouldWrapLabels = ["ar", "bg", "el", "en", "ko", "zh-HK", "zh-TW"].includes(this.messages._t9nLocale);
  }
  render() {
    const e = this.mode === "single" && !this.disabled;
    return k(this.messages._lang, d`<arcgis-slider .ariaLabel=${this.label} class=${_({ "wrap-labels": this._shouldWrapLabels })} .disabled=${this.disabled} fill-placement=none .labelFormatter=${this._sliderLabelFormatter} layout=horizontal min=0 .max=${o} .steps=${this.steps ?? 1} .valueLabelsDisplay=${this.mode === "range" ? "hidden" : "always"} value-labels-placement=start .values=${this._values} @arcgisInput=${({ currentTarget: { values: t } }) => {
      this.mode === "single" ? this._setStartTime(t[0]) : this._setRange(t[0], t[1]);
    }}>${this.mode === "range" ? this._renderTimeRange() : void 0}<div class="ticks-container" slot=content-end><arcgis-ticks class="main-ticks" .interactive=${e} .labelFormatter=${this._tickLabelFormatter} layout=horizontal min=0 .max=${o} mode=value show-labels .values=${g} @arcgisTickClick=${(t) => this._setStartTime(t.detail.value)}></arcgis-ticks><arcgis-ticks class="secondary-ticks" layout=horizontal min=0 .max=${o} mode=value .values=${$} @arcgisTickClick=${(t) => this._setStartTime(t.detail.value)}></arcgis-ticks></div></arcgis-slider>`);
  }
  _renderTimeRange() {
    const e = this._sliderLabelFormatter(this.startTime), t = this._sliderLabelFormatter(this.endTime), a = this.messages?.timeRange;
    return d`<div class="time-range-label" aria-live=polite slot=content-start>${T(a ?? "{start} — {end}", { start: e, end: t })}</div>`;
  }
}
const o = i(24, "hours", "minutes") - 1, g = [0, 6, 12, 18].map((l) => i(l, "hours", "minutes"));
g.push(o);
const $ = [2, 4, 8, 10, 14, 16, 20, 22].map((l) => i(l, "hours", "minutes")), h = {
  hour: "numeric",
  minute: "numeric",
  timeZone: "UTC"
}, y = {
  hour: "numeric",
  timeZone: "UTC"
};
u("arcgis-time-of-day-slider", z);
export {
  z as ArcgisTimeOfDaySlider
};
