// @arcgis/map-components/dist/components/arcgis-compass/customElement.js
/* COPYRIGHT Esri - https://js.arcgis.com/5.0/LICENSE.txt */
import { c as s } from "../../chunks/runtime.js";
import { css as a, html as r } from "lit";
import c from "@arcgis/core/widgets/Compass/CompassViewModel.js";
import { LitElement as l, createEvent as n, safeClassMap as h, nothing as i, safeStyleMap as d } from "@arcgis/lumina";
import { usePropertyChange as v } from "@arcgis/lumina/controllers";
import { classes as m } from "@arcgis/toolkit/dom";
import { u as g } from "../../chunks/useT9n.js";
import { m as u } from "../../chunks/useViewModel.js";
import { g as p } from "../../chunks/globalCss.js";
const f = a`.arcgis-button{--calcite-icon-color: var(--calcite-color-text-3);--calcite-button-background-color: var(--calcite-color-foreground-1);color:var(--calcite-color-text-3);line-height:var(--calcite-font-line-height-relative-tight);display:flex;flex-flow:column;cursor:pointer;background-color:var(--calcite-color-foreground-1);font-size:var(--calcite-font-size);:hover{--calcite-icon-color: var(--calcite-color-text-1);--calcite-button-background-color: var(--calcite-color-transparent-hover)}}:host{calcite-button{width:var(--calcite-spacing-xxxl);height:var(--calcite-spacing-xxxl)}}:host([visual-scale="l"]){calcite-button{width:calc(var(--calcite-spacing-xxxl) + var(--calcite-spacing-md));height:calc(var(--calcite-spacing-xxxl) + var(--calcite-spacing-md))}}:host([visual-scale="s"]){calcite-button{width:var(--calcite-spacing-xxl);height:var(--calcite-spacing-xxl)}}.root{border-radius:50%;box-shadow:var(--arcgis-internal-box-shadow, none)}.icon-container{justify-content:center;display:flex;position:absolute;transform:translate(-50%,-50%)}.icon-container:dir(rtl){transform:translate(50%,-50%)}`, x = u(c);
class b extends l {
  constructor() {
    super(...arguments), this.messages = g(), this.viewModel = x(this), this.view = this.viewModel.view, this.autoDestroyDisabled = !1, this.goToOverride = this.viewModel.goToOverride, this.orientation = this.viewModel.orientation, this.state = this.viewModel.state, this.visualScale = "m", this.arcgisPropertyChange = v()("state", "orientation"), this.arcgisReady = n();
  }
  static {
    this.properties = { view: 0, autoDestroyDisabled: 5, goToOverride: 0, icon: 3, label: 1, messageOverrides: 0, orientation: 32, referenceElement: 1, state: 32, visualScale: 3 };
  }
  static {
    this.styles = f;
  }
  get icon() {
    return this._icon ?? (this.state === "rotation" ? "arrow-up" : "compass-needle");
  }
  set icon(t) {
    this._icon = t;
  }
  async destroy() {
    await this.manager.destroy();
  }
  async reset() {
    this.viewModel.reset();
  }
  _reset() {
    this.viewModel.reset();
  }
  render() {
    const { messages: t, icon: o } = this;
    return r`<div class=${h(m("root", p.arcgisButton))}><calcite-button .disabled=${this.state === "disabled"} kind=neutral .label=${t.reset} @click=${this._reset} round .scale=${this.visualScale} title=${t.reset ?? i}><div aria-hidden=true class="icon-container" title=${t.reset ?? i}><calcite-icon .scale=${this.visualScale} .icon=${o} style=${d(w(this.orientation))}></calcite-icon></div></calcite-button></div>`;
  }
}
const w = (e) => ({
  transform: `rotateZ(${e.z}deg)`
});
s("arcgis-compass", b);
export {
  b as ArcgisCompass
};
