// @arcgis/map-components/dist/components/arcgis-print-map-only-advanced-options/customElement.js
/* COPYRIGHT Esri - https://js.arcgis.com/5.0/LICENSE.txt */
import { c as i } from "../../chunks/runtime.js";
import { css as s, html as e } from "lit";
import { LitElement as n } from "@arcgis/lumina";
const r = s`.advanced-options-container{padding:var(--calcite-spacing-xxs);display:flex;flex-direction:column;gap:var(--calcite-spacing-xs)}`;
class a extends n {
  static {
    this.properties = { viewModel: 0, messages: 0 };
  }
  static {
    this.styles = r;
  }
  loaded() {
    this.manager.onLifecycle(() => []);
  }
  render() {
    return e`<div aria-labelledby=${`${this.el.id}__advancedOptionsForLayout`} class="advanced-options-container">${this._renderScaleSection()}${this._renderDpiSection()}${this._renderOutSpatialReferenceSection()}${this._renderAttributionSection()}${this._renderCustomTextElementSection()}</div>`;
  }
  _renderScaleSection() {
    return e`<arcgis-print-scale-section .messages=${this.messages} .viewModel=${this.viewModel}></arcgis-print-scale-section>`;
  }
  _renderDpiSection() {
    return e`<arcgis-print-dpi-input .messages=${this.messages} .viewModel=${this.viewModel}></arcgis-print-dpi-input>`;
  }
  _renderOutSpatialReferenceSection() {
    return e`<arcgis-print-out-spatial-reference-input .messages=${this.messages} .viewModel=${this.viewModel}></arcgis-print-out-spatial-reference-input>`;
  }
  _renderAttributionSection() {
    return e`<div><calcite-label layout=inline><calcite-checkbox .checked=${this.viewModel.templateOptions.attributionEnabled} @calciteCheckboxChange=${(t) => this.viewModel.templateOptions.attributionEnabled = t.target.checked}></calcite-checkbox>${this.messages.attribution}</calcite-label></div>`;
  }
  _renderCustomTextElementSection() {
    return e`<arcgis-print-custom-text-elements .viewModel=${this.viewModel}></arcgis-print-custom-text-elements>`;
  }
}
i("arcgis-print-map-only-advanced-options", a);
export {
  a as PrintMapOnlyAdvancedOptions
};
