// @arcgis/map-components/dist/components/arcgis-presentation/customElement.js
/* COPYRIGHT Esri - https://js.arcgis.com/5.0/LICENSE.txt */
import { c as v } from "../../chunks/runtime.js";
import { css as g, html as c } from "lit";
import { useDirection as u } from "@arcgis/lumina/controllers";
import { LitElement as S, createEvent as m, nothing as h } from "@arcgis/lumina";
import { u as b } from "../../chunks/useT9n.js";
import { m as w } from "../../chunks/useViewModel.js";
import { property as d, subclass as y } from "@arcgis/core/core/accessorSupport/decorators.js";
import f from "@arcgis/core/core/Accessor.js";
import p from "@arcgis/core/core/Collection.js";
import { ignoreAbortErrors as x } from "@arcgis/core/core/promiseUtils.js";
import { on as I, watch as _ } from "@arcgis/core/core/reactiveUtils.js";
const E = g`:host{width:100%;height:100%;padding:var(--calcite-spacing-none);margin:var(--calcite-spacing-none);display:flex}p{padding:0;margin:0}.root{--internal-attribution-height: 16px;pointer-events:auto;display:flex;align-items:center;gap:var(--calcite-spacing-sm);background-color:var(--calcite-color-foreground-1);box-shadow:var(--arcgis-internal-box-shadow);align-self:flex-end;margin-block:auto calc(var(--calcite-spacing-fixed-xl) + var(--internal-attribution-height));margin-inline-start:var(--calcite-spacing-fixed-xl)}.slide-counter-text{display:flex;align-items:center;justify-content:center;gap:var(--calcite-spacing-base);min-width:64px}.slide-counter-text-element{box-sizing:border-box;max-width:28px;margin:0;padding:0;overflow:hidden;white-space:nowrap;text-overflow:ellipsis;font-variant-numeric:tabular-nums}`;
var M = Object.defineProperty, A = Object.getOwnPropertyDescriptor, r = (e, t, i, l) => {
  for (var n = l > 1 ? void 0 : l ? A(t, i) : t, a = e.length - 1, o; a >= 0; a--)
    (o = e[a]) && (n = (l ? o(t, i, n) : o(n)) || n);
  return l && n && M(t, i, n), n;
};
let s = class extends f {
  //#region Lifecycle
  constructor(e) {
    super(e), this.keyboardNavigationEnabled = !1, this._subscribeToSlideEvents = (t) => {
      this._unSubscribeToSlideEvents(t), t.forEach((i) => {
        const l = i.on("apply-slide-start", () => this._onApplySlideStart(i));
        this.addHandles(l, i.id);
      });
    }, this._unSubscribeToSlideEvents = (t) => {
      t.forEach((i) => {
        this.hasHandles(i.id) && this.removeHandles(i.id);
      });
    };
  }
  initialize() {
    this._startWatchVisibleSlides();
  }
  destroy() {
    this._unSubscribeToSlideEvents(this.visibleSlides);
  }
  get visibleSlides() {
    return this.view?.map?.presentation?.slides?.filter((e) => !e.hidden) ?? new p();
  }
  get lastSlideSelected() {
    return this.getSlideIndexBySlideId(this.slideId) === this.visibleSlides.length - 1;
  }
  get firstSlideSelected() {
    return this.getSlideIndexBySlideId(this.slideId) === 0;
  }
  get slideSelected() {
    return this.getSlideIndexBySlideId(this.slideId) > -1;
  }
  get hasOnlyOneSlide() {
    return this.visibleSlides.length === 1;
  }
  get slideIndex() {
    return this.getSlideIndexBySlideId(this.slideId);
  }
  getSlideIndexBySlideId(e) {
    let t = -1;
    return e && (t = this.visibleSlides.findIndex((i) => i.id === e)), t;
  }
  navigateInPresentation(e) {
    const t = this.slideIndex + e;
    let i = this._getSlideBySlideIndex(t);
    i ??= this._getSlideBySlideIndex(0), i && this._applySlide(i);
  }
  captureArrowKeysForNavigation(e, t) {
    const i = ["ArrowLeft", "ArrowRight", "PageUp", "PageDown"];
    if (document.activeElement?.nodeName.toLowerCase() === "arcgis-scene" || !i.includes(e.code))
      return;
    const n = t === "rtl" ? "ArrowRight" : "ArrowLeft", o = e.code === n || e.code === "PageUp" ? -1 : 1;
    this.navigateInPresentation(o);
  }
  //#endregion
  //#region Private Methods
  _getSlideBySlideIndex(e) {
    if (e != null && e >= 0 && e < this.visibleSlides.length)
      return this.visibleSlides.getItemAt(e);
  }
  async _applySlide(e) {
    this.view && await x(e.applyTo(this.view));
  }
  _startWatchVisibleSlides() {
    this._subscribeToSlideEvents(this.visibleSlides), this.addHandles(
      I(
        () => this.visibleSlides,
        "change",
        (e) => {
          this._subscribeToSlideEvents(new p(e.added)), this._unSubscribeToSlideEvents(new p(e.removed));
        },
        {
          onListenerAdd: this._subscribeToSlideEvents,
          onListenerRemove: this._unSubscribeToSlideEvents
        }
      )
    );
  }
  _onApplySlideStart(e) {
    this.slideId = e.id;
  }
  //#endregion
};
r([
  d()
], s.prototype, "view", 2);
r([
  d()
], s.prototype, "slideId", 2);
r([
  d()
], s.prototype, "keyboardNavigationEnabled", 2);
r([
  d()
], s.prototype, "visibleSlides", 1);
r([
  d()
], s.prototype, "lastSlideSelected", 1);
r([
  d()
], s.prototype, "firstSlideSelected", 1);
r([
  d()
], s.prototype, "slideSelected", 1);
r([
  d()
], s.prototype, "hasOnlyOneSlide", 1);
r([
  d()
], s.prototype, "getSlideIndexBySlideId", 1);
s = r([
  y("components.presentation.PresentationViewModel")
], s);
const $ = w(s);
class P extends S {
  constructor() {
    super(...arguments), this._messages = b(), this.viewModel = $(this), this._direction = u(), this._captureArrowKeysForNavigation = (t) => {
      this.viewModel.captureArrowKeysForNavigation(t, this._direction);
    }, this.autoDestroyDisabled = !1, this.view = this.viewModel.view, this.scale = "m", this.keyboardNavigationEnabled = !1, this.arcgisReady = m();
  }
  static {
    this.properties = { slideId: 1, autoDestroyDisabled: 5, view: 0, scale: 1, keyboardNavigationEnabled: 5 };
  }
  static {
    this.styles = E;
  }
  get slideId() {
    return this.viewModel.slideId;
  }
  set slideId(t) {
    this.viewModel.slideId = t;
  }
  async destroy() {
    await this.manager.destroy();
  }
  loaded() {
    this.manager.onLifecycle(() => _(() => this.keyboardNavigationEnabled, (i) => {
      document.removeEventListener("keydown", this._captureArrowKeysForNavigation), i && document.addEventListener("keydown", this._captureArrowKeysForNavigation);
    }, { initial: !0 })), this.manager.onDisconnected(() => {
      document.removeEventListener("keydown", this._captureArrowKeysForNavigation);
    });
  }
  render() {
    const t = this.viewModel.lastSlideSelected && !this.viewModel.hasOnlyOneSlide, i = this._direction === "rtl", l = i ? "chevron-right" : "chevron-left", n = i ? "chevron-left" : "chevron-right";
    return this.viewModel.visibleSlides.length === 0 ? c`<div></div>` : c`<div class="root"><calcite-button appearance=transparent icon-start=${l} kind=neutral .scale=${this.scale} .disabled=${this.viewModel.firstSlideSelected || !this.viewModel.slideSelected} .label=${this._messages.previous} title=${this._messages.previous ?? h} @click=${() => this.viewModel.navigateInPresentation(-1)}></calcite-button><div class="slide-counter-text">${this._renderSlideCounter()}</div><calcite-button appearance=transparent icon-start=${t ? "reset" : n} kind=neutral .scale=${this.scale} .label=${t ? this._messages.restart : this._messages.next} title=${(t ? this._messages.restart : this._messages.next) ?? h} @click=${() => this.viewModel.navigateInPresentation(1)}></calcite-button></div>`;
  }
  _renderSlideCounter() {
    const t = this.viewModel.slideIndex + 1, i = this._messages?.slideNavAriaText?.replace("${index}", t.toString()).replace("${numberOfSlides}", this.viewModel.visibleSlides.length.toString());
    return this.viewModel.slideSelected ? c`<p role=banner aria-label=${i ?? h} aria-live=polite><div class="slide-counter-text-element" data-testId=arcgis-map-presentation-slideIndex>${t}</div><div class="slide-counter-text-element">/</div><p class="slide-counter-text-element" data-testId=arcgis-map-presentation-numberOfSlides>${this.viewModel.visibleSlides.length}</p></p>` : c`<p role=banner data-testId=arcgis-map-presentation-numberOfSlides>${this.viewModel.visibleSlides?.length} ${this._messages?.slides}</p>`;
  }
}
v("arcgis-presentation", P);
export {
  P as ArcgisPresentation
};
